# Load dependencies

In [1]:
import cobra
import libsbml
import lxml

In [2]:
models_directory = '/media/sf_Shared/Systems_biology/Metabolic_models/'
OB3b_directory = '/media/sf_Shared/GEM_OB3b/'
memote_directory = '/home/ensakz/Desktop/memote_m_trichosporium/'
fastas_directory = '/media/sf_Shared/Systems_biology/Fastas_and_annotations/'
inparanoid_directory = '/media/sf_Shared/Systems_biology/InParanoid_runs/'

In [69]:
old_models_directory = "/media/sf_Shared/GEM_OB3b/Previous_work_2018/"+\
"OB3b_history_files_from_ubuntu_virtual_box_desktop/"

In [720]:
AM1_bigg = cobra.io.read_sbml_model(old_models_directory + "AM1_model_methylotrophic_network_BIGG_"+\
                                    "renamedGPR_renamed_workable_model_only_essential_reactions.xml")

In [72]:
model_Bath = cobra.io.load_json_model(models_directory + "iMcBath.json")

In [114]:
AM1_OB3b_combined=cobra.io.read_sbml_model(models_directory + "OB3b_draft_GEM_based_on_AM1_and_Bath_GEMs.xml")    

## AM1_OB3b old model BiGG exploration:

Loading combined AM1 GEM with OB3b:

In [76]:
META_reactions = []
for reaction in AM1_bigg.reactions:
    if "META" in reaction.gene_reaction_rule:
        META_reactions.append(reaction)

In [77]:
print(len(META_reactions))
print(META_reactions)

60
[<Reaction EAOXRED at 0x7fd7deac5e10>, <Reaction PPND at 0x7fd7deae4390>, <Reaction MNEUROPNDH3 at 0x7fd7deae4908>, <Reaction MNEUROPNDH2 at 0x7fd7deae4c50>, <Reaction PEPGLYCANGT at 0x7fd7deaf23c8>, <Reaction THRD_L at 0x7fd7deaf24e0>, <Reaction EDTXS1 at 0x7fd7de8ce908>, <Reaction PRE3BS at 0x7fd7de8ce940>, <Reaction SULabcpp at 0x7fd7de8e97f0>, <Reaction PC6YM at 0x7fd7de8fb438>, <Reaction MOBDMPTL at 0x7fd7de8492e8>, <Reaction LYSTRS at 0x7fd7de870a20>, <Reaction COBALT2tpp at 0x7fd7de7dc400>, <Reaction RNDR3 at 0x7fd7de7dccc0>, <Reaction RNDR2 at 0x7fd7de7dcda0>, <Reaction RNDR1 at 0x7fd7de7ed7b8>, <Reaction ALATA_D at 0x7fd7de70e860>, <Reaction ZCARDS2 at 0x7fd7de70ecc0>, <Reaction CHORM at 0x7fd7de6c5f60>, <Reaction ZCARDS at 0x7fd7de6e2be0>, <Reaction MOGDS at 0x7fd7de6f6cf8>, <Reaction CPPPGO at 0x7fd7de647c50>, <Reaction FRDOr at 0x7fd7de67dd30>, <Reaction ORPT at 0x7fd7de443e48>, <Reaction FMNRx2 at 0x7fd7de468c18>, <Reaction ORNDC at 0x7fd7de2992e8>, <Reaction HNEUROPNS2

In [78]:
AM1_bigg.reactions

[<Reaction EAOXRED at 0x7fd7deac5e10>,
 <Reaction IMPC at 0x7fd7dead4358>,
 <Reaction HACD8i at 0x7fd7dead44e0>,
 <Reaction PANTS at 0x7fd7dead4940>,
 <Reaction IMPD at 0x7fd7dead4f60>,
 <Reaction OCPPH3 at 0x7fd7dead9400>,
 <Reaction PGSA161 at 0x7fd7dead97b8>,
 <Reaction LCAD at 0x7fd7dead9a58>,
 <Reaction PGSA160 at 0x7fd7dead9c18>,
 <Reaction GLUDy at 0x7fd7dead9e10>,
 <Reaction PPAtex at 0x7fd7deae40b8>,
 <Reaction DTMPK at 0x7fd7deae40f0>,
 <Reaction PPND at 0x7fd7deae4390>,
 <Reaction DXPRIi at 0x7fd7deae44a8>,
 <Reaction PC11M at 0x7fd7deae4710>,
 <Reaction PSERTr at 0x7fd7deae4828>,
 <Reaction CAT at 0x7fd7deae4a20>,
 <Reaction CYO_2p at 0x7fd7deae4a90>,
 <Reaction UAMAS at 0x7fd7deae4da0>,
 <Reaction MNEUROPNDH3 at 0x7fd7deae4908>,
 <Reaction EX_DM_trnathr at 0x7fd7deae4b70>,
 <Reaction MNEUROPNDH2 at 0x7fd7deae4c50>,
 <Reaction 2AGPEAT161acp at 0x7fd7deae4e10>,
 <Reaction PEPGLYCANGT at 0x7fd7deaf23c8>,
 <Reaction MTHFC at 0x7fd7deaf25c0>,
 <Reaction MTHFD at 0x7fd7deaf27b8>

### Add methane exchange and 3 modes of methane oxidation to the GEM:

In [723]:
reaction_1 = model_Bath.reactions.EX_ch4_e
reaction_2 = model_Bath.reactions.CH4tex
reaction_3 = model_Bath.reactions.PMMOipp #redox-arm, change cofactor
reaction_4 = model_Bath.reactions.PMMODCipp #direct coupling, change cofactor
#reaction_5 = model_Bath.reactions.CYOR_q8ppi #uphill electron transfer, don't add this reaction, change cofactor 
                                             #in ubiquinol-cytochrome-c reductase and keep its reversible and keep
                                             #pMMOipp active

reaction_3 manipulation:

In [724]:
reaction_3

<Reaction PMMOipp at 0x7fd7df00f320>

In [725]:
reaction_3.reaction = "ch4_p + o2_p + qh2_c --> h2o_p + meoh_p + q_c"

unknown metabolite 'qh2_c' created
unknown metabolite 'q_c' created


reaction_4 manipulation:

In [726]:
reaction_4

<Reaction PMMODCipp at 0x7fd7df033470>

In [727]:
reaction_4.reaction = "ch4_p + 2.0 focytc_c + 2.0 h_p + o2_p --> 2.0 ficytc_c + h2o_p + meoh_p"

unknown metabolite 'focytc_c' created
unknown metabolite 'ficytc_c' created


reaction_5 adjustment:

In [728]:
reaction_5

NameError: name 'reaction_5' is not defined

In [729]:
for reaction in AM1_bigg.metabolites.q_c.reactions:
    print(reaction.name, reaction)

succinate_dehydrogenase SUCCD1: q_c + succ_c <=> fum_c + qh2_c
ETF_dehydrogenase ETFDH: q_c + rETF_c <=> oxETF_c + qh2_c
cytochrome_o_ubiquinol_oxidase CYOO_4p: 4.0 h_c + 0.5 o2_c + qh2_c --> h2o_c + 4.0 h_p + q_c
cytochrome_d_terminal_oxidase CYOD: 4.0 h_c + o2_c + 2.0 qh2_c --> 2.0 h2o_c + 4.0 h_p + 2.0 q_c
NADH_ubiquinone_oxidoreductase_I NADH11: 5.0 h_c + nadh_c + q_c --> 4.0 h_p + nad_c + qh2_c
nitrate_reductase__periplasmic_ NO3Rpp: no3_p + qh2_c --> h2o_p + no2_p + q_c
ubiquinol_cytochrome_c_reductase UQCYOR_2p: 2.0 ficytc_c + qh2_c --> 2.0 focytc_c + 2.0 h_p + q_c
photosystem_I_cyclic PHOTOSYS1C: 2.0 fdxrd_c + 2.0 h_c + q_c --> 2.0 fdxox_c + qh2_c


In [730]:
AM1_bigg.reactions.UQCYOR_2p

<Reaction UQCYOR_2p at 0x7fd7d29365c0>

Add reactions 1-4 to AM1_bigg GEM:

In [731]:
methane_reactions_to_add = [reaction_1, reaction_2, reaction_3, reaction_4]

In [732]:
for reaction in methane_reactions_to_add:
    print(reaction.id, reaction.reaction)

EX_ch4_e ch4_e <-- 
CH4tex ch4_e <=> ch4_p
PMMOipp ch4_p + o2_p + qh2_c --> h2o_p + meoh_p + q_c
PMMODCipp ch4_p + 2.0 focytc_c + 2.0 h_p + o2_p --> 2.0 ficytc_c + h2o_p + meoh_p


In [733]:
AM1_bigg.add_reactions(methane_reactions_to_add)

In [90]:
AM1_bigg.summary()

IN FLUXES             OUT FLUXES     OBJECTIVES
--------------------  -------------  ----------------------
o2_e       26.8       h2o_e  36.6    Biomass_Mext...  0.355
ch4_e      18.5       co2_e   7.21
nh4_e       3.47      h_e     0.791
meoh_e      3
pi_e        0.248
k_e         0.0654
so4_e       0.0224
mg2_e       0.00292
fe2_e       0.00285
cl_e        0.00175
ca2_e       0.00175
na1_e       0.00144
mobd_e      0.00125
cu2_e       0.00117
mn2_e       0.00117
zn2_e       0.00117
cobalt2_e   7.79e-05


Let's knockout meoh_e first:

In [91]:
AM1_bigg.reactions.EX_meoh_e.knock_out()

In [92]:
AM1_bigg.summary()

IN FLUXES             OUT FLUXES     OBJECTIVES
--------------------  -------------  ----------------------
o2_e       24.7       h2o_e  31.7    Biomass_Mext...  0.297
ch4_e      18.5       co2_e   6.54
nh4_e       2.9       h_e     0.662
pi_e        0.207
k_e         0.0547
so4_e       0.0187
mg2_e       0.00245
fe2_e       0.00238
cl_e        0.00147
ca2_e       0.00147
na1_e       0.00121
mobd_e      0.00104
cu2_e       0.000979
mn2_e       0.000979
zn2_e       0.000979
cobalt2_e   6.52e-05


#### Let's knockout PMMODCipp second, so that only PMMOipp (redox arm is active):

In [93]:
AM1_bigg.reactions.PMMODCipp.knock_out()

In [94]:
AM1_bigg.summary()

IN FLUXES             OUT FLUXES    OBJECTIVES
--------------------  ------------  ----------------------
o2_e       27.7       h2o_e  33     Biomass_Mext...  0.224
ch4_e      18.5       co2_e   9.46
nh4_e       2.19      h_e     0.5
pi_e        0.157
k_e         0.0413
so4_e       0.0141
mg2_e       0.00185
fe2_e       0.0018
ca2_e       0.00111
cl_e        0.00111
na1_e       0.000912
mobd_e      0.000788
cu2_e       0.000739
mn2_e       0.000739
zn2_e       0.000739
cobalt2_e   4.92e-05


Try to change reaction bounds to lower methane uptake rates:

In [734]:
AM1_bigg.reactions.EX_ch4_e.bounds = (-4.079, -4.079)

In [171]:
with AM1_bigg as model:
    reaction_id = "Biomass_Mextorquens_AM1_core"
    model.objective = reaction_id
    model.optimize()
    model_objective = model.optimize().fluxes
    model_objective.to_csv(Escher_gapfilling_dir + reaction_id + "_PMMOipp"+ "_AM1" + '.csv')
    model.summary()

IN FLUXES            OUT FLUXES    OBJECTIVES
-------------------  ------------  -----------------------
o2_e       6.12      h2o_e  7.28   Biomass_Mext...  0.0495
ch4_e      4.08      co2_e  2.09
nh4_e      0.484     h_e    0.11
pi_e       0.0346
k_e        0.00913
so4_e      0.00312
mg2_e      0.000408
fe2_e      0.000397
ca2_e      0.000245
cl_e       0.000245
na1_e      0.000201
mobd_e     0.000174
cu2_e      0.000163
mn2_e      0.000163
zn2_e      0.000163
cobalt2_e  1.09e-05


In [174]:
AM1_bigg.metabolites.o2_c.summary()

PRODUCING REACTIONS -- O2 (o2_c)
--------------------------------
%       FLUX  RXN ID    REACTION
----  ------  --------  --------------------------------------------------
100%    2.04  O2tpp     o2_p <=> o2_c

CONSUMING REACTIONS -- O2 (o2_c)
--------------------------------
%       FLUX  RXN ID    REACTION
----  ------  --------  --------------------------------------------------
100%    2.04  CYO_2p    2.0 focytc_c + 4.0 h_c + 0.5 o2_c --> 2.0 ficyt...


In [177]:
model_Bath.reactions.CyLCyHORpp

<Reaction CyLCyHORpp at 0x7fd7df2312b0>

In [178]:
model_Bath.reactions.MEOHDHipp

<Reaction MEOHDHipp at 0x7fd7deb29f28>

This looks quite similar to Bath GEM

#### Let's try direct coupling:
Direct coupling between methane monooxygenase and methanol dehydrogenase

In [97]:
with AM1_bigg:
    AM1_bigg.reactions.PMMOipp.knock_out()
    AM1_bigg.reactions.PMMODCipp.bounds = (0, 1000)
    AM1_bigg.summary()

IN FLUXES            OUT FLUXES    OBJECTIVES
-------------------  ------------  -----------------------
o2_e       5.46      h2o_e  7      Biomass_Mext...  0.0656
ch4_e      4.08      co2_e  1.44
nh4_e      0.641     h_e    0.146
pi_e       0.0458
k_e        0.0121
so4_e      0.00414
mg2_e      0.000541
fe2_e      0.000526
ca2_e      0.000325
cl_e       0.000325
na1_e      0.000267
mobd_e     0.000231
cu2_e      0.000216
mn2_e      0.000216
zn2_e      0.000216
cobalt2_e  1.44e-05


In [179]:
with AM1_bigg as model:
    reaction_id = "Biomass_Mextorquens_AM1_core"
    model.objective = reaction_id
    model.reactions.PMMOipp.knock_out()
    model.reactions.PMMODCipp.bounds = (0, 1000)
    model.optimize()
    model_objective = model.optimize().fluxes
    model_objective.to_csv(Escher_gapfilling_dir + reaction_id + "_PMMODCipp"+ "_AM1" + '.csv')
    model.summary()

IN FLUXES            OUT FLUXES    OBJECTIVES
-------------------  ------------  -----------------------
o2_e       5.46      h2o_e  7      Biomass_Mext...  0.0656
ch4_e      4.08      co2_e  1.44
nh4_e      0.641     h_e    0.146
pi_e       0.0458
k_e        0.0121
so4_e      0.00414
mg2_e      0.000541
fe2_e      0.000526
ca2_e      0.000325
cl_e       0.000325
na1_e      0.000267
mobd_e     0.000231
cu2_e      0.000216
mn2_e      0.000216
zn2_e      0.000216
cobalt2_e  1.44e-05


In [183]:
AM1_bigg.reactions.ATPS4rpp

<Reaction ATPS4rpp at 0x7fd7dd507e10>

May be my issue is in complex IV where I can regenerate qh2_c using complex IV variants. In Bath GEM, I cannot do it. --> Check electron transport chains of other methanotrophs. 

Let's try Methylococcus capsulatus Bath implementation for direct coupling:

In [186]:
with AM1_bigg as model:
    reaction_id = "Biomass_Mextorquens_AM1_core"
    model.objective = reaction_id
    model.reactions.PMMOipp.knock_out()
    model.reactions.CYOD.knock_out()
    model.reactions.CYOO_4p.knock_out()
    model.reactions.PMMODCipp.bounds = (0, 1000)
    model.optimize()
    model_objective = model.optimize().fluxes
    model_objective.to_csv(Escher_gapfilling_dir + reaction_id + "_PMMODCipp"+ "_AM1" + '.csv')
    model.summary()

IN FLUXES            OUT FLUXES    OBJECTIVES
-------------------  ------------  -----------------------
o2_e       5.46      h2o_e  7      Biomass_Mext...  0.0656
ch4_e      4.08      co2_e  1.44
nh4_e      0.641     h_e    0.146
pi_e       0.0458
k_e        0.0121
so4_e      0.00414
mg2_e      0.000541
fe2_e      0.000526
ca2_e      0.000325
cl_e       0.000325
na1_e      0.000267
mobd_e     0.000231
cu2_e      0.000216
mn2_e      0.000216
zn2_e      0.000216
cobalt2_e  1.44e-05


Save your GEM:

In [348]:
cobra.io.save_json_model(AM1_bigg, memote_directory + "AM1_bigg.json")

Sanity check before going further:

In [98]:
with AM1_bigg:
    AM1_bigg.reactions.PMMOipp.knock_out()
    AM1_bigg.reactions.PMMODCipp.bounds = (0, 1000)
    #AM1_bigg.reactions.UQCYOR_2p.knock_out()
    AM1_bigg.summary()
    AM1_bigg.metabolites.q_c.summary()

IN FLUXES            OUT FLUXES    OBJECTIVES
-------------------  ------------  -----------------------
o2_e       5.46      h2o_e  7      Biomass_Mext...  0.0656
ch4_e      4.08      co2_e  1.44
nh4_e      0.641     h_e    0.146
pi_e       0.0458
k_e        0.0121
so4_e      0.00414
mg2_e      0.000541
fe2_e      0.000526
ca2_e      0.000325
cl_e       0.000325
na1_e      0.000267
mobd_e     0.000231
cu2_e      0.000216
mn2_e      0.000216
zn2_e      0.000216
cobalt2_e  1.44e-05
PRODUCING REACTIONS -- UQ (q_c)
-------------------------------
%       FLUX  RXN ID    REACTION
----  ------  --------  --------------------------------------------------
100%   2.76   CYOO_4p   4.0 h_c + 0.5 o2_c + qh2_c --> h2o_c + 4.0 h_p ...

CONSUMING REACTIONS -- UQ (q_c)
-------------------------------
%       FLUX  RXN ID    REACTION
----  ------  --------  --------------------------------------------------
72%    1.98   NADH11    5.0 h_c + nadh_c + q_c --> 4.0 h_p + nad_c + qh2_c
14%    0.387  ETFDH

In [99]:
AM1_bigg.metabolites.q_c.summary()

PRODUCING REACTIONS -- UQ (q_c)
-------------------------------
%       FLUX  RXN ID    REACTION
----  ------  --------  --------------------------------------------------
100%   4.08   PMMOipp   ch4_p + o2_p + qh2_c --> h2o_p + meoh_p + q_c

CONSUMING REACTIONS -- UQ (q_c)
-------------------------------
%       FLUX  RXN ID    REACTION
----  ------  --------  --------------------------------------------------
48%    1.95   SUCCD1    q_c + succ_c <=> fum_c + qh2_c
45%    1.83   NADH11    5.0 h_c + nadh_c + q_c --> 4.0 h_p + nad_c + qh2_c
7%     0.292  ETFDH     q_c + rETF_c <=> oxETF_c + qh2_c


In [100]:
with AM1_bigg:
    #AM1_bigg.reactions.PMMOipp.knock_out()
    #AM1_bigg.reactions.PMMODCipp.bounds = (0, 1000)
    AM1_bigg.reactions.UQCYOR_2p.bounds = (-1000, 1000)
    AM1_bigg.summary()
    AM1_bigg.metabolites.q_c.summary()

IN FLUXES            OUT FLUXES    OBJECTIVES
-------------------  ------------  -----------------------
o2_e       5.46      h2o_e  7      Biomass_Mext...  0.0656
ch4_e      4.08      co2_e  1.44
nh4_e      0.641     h_e    0.146
pi_e       0.0458
k_e        0.0121
so4_e      0.00414
mg2_e      0.000541
fe2_e      0.000526
ca2_e      0.000325
cl_e       0.000325
na1_e      0.000267
mobd_e     0.000231
cu2_e      0.000216
mn2_e      0.000216
zn2_e      0.000216
cobalt2_e  1.44e-05
PRODUCING REACTIONS -- UQ (q_c)
-------------------------------
%       FLUX  RXN ID     REACTION
----  ------  ---------  --------------------------------------------------
100%   4.08   PMMOipp    ch4_p + o2_p + qh2_c --> h2o_p + meoh_p + q_c

CONSUMING REACTIONS -- UQ (q_c)
-------------------------------
%       FLUX  RXN ID     REACTION
----  ------  ---------  --------------------------------------------------
49%    1.98   NADH11     5.0 h_c + nadh_c + q_c --> 4.0 h_p + nad_c + qh2_c
32%    1.32   UQCY

In [101]:
AM1_bigg.metabolites.h2o_p.summary()

PRODUCING REACTIONS -- H2O (h2o_p)
----------------------------------
%       FLUX  RXN ID    REACTION
----  ------  --------  ---------------------------------------------
56%     4.08  PMMOipp   ch4_p + o2_p + qh2_c --> h2o_p + meoh_p + q_c
44%     3.2   H2Otpp    h2o_p <=> h2o_c

CONSUMING REACTIONS -- H2O (h2o_p)
----------------------------------
%       FLUX  RXN ID    REACTION
----  ------  --------  ---------------------------------------------
100%    7.28  H2Otex    h2o_e <=> h2o_p


In [102]:
AM1_bigg.reactions.UQCYOR_2p

<Reaction UQCYOR_2p at 0x7fd7de61a6d8>

In [103]:
AM1_bigg.reactions.CYO_2p

<Reaction CYO_2p at 0x7fd7deae4a90>

#### Let's try uphill electron transfer:

In [104]:
with AM1_bigg:
    AM1_bigg.reactions.PMMODCipp.knock_out()
    AM1_bigg.reactions.PMMOipp.bounds = (0, 1000)
    AM1_bigg.reactions.UQCYOR_2p.bounds = (-1000, 1000)

It seems like there is no difference in growth rate predictions between direct coupling and uphill electron flow modes of methane oxidation:
You have 2 options to understand this:
1. Draw or restructure Escher map for OB3b central carbon metabolism.
2. Understand modes of methane oxidation in detail. Refer to specific literature, Bath GEM Escher map and reconstruction. Try to correct these problems

Save GEM in a json format so that it can loaded into Escher. --> I have already saved GEM in json format yesterday

Look for flux in metabolites so that you can draw Escher GEM reconstruction:

In [377]:
AM1_bigg.metabolites.fald_c.summary()

PRODUCING REACTIONS -- formaldehyde (fald_c)
--------------------------------------------
%       FLUX  RXN ID    REACTION
----  ------  --------  ----------------------------------
100%    4.08  FALDtpp   fald_p <=> fald_c

CONSUMING REACTIONS -- formaldehyde (fald_c)
--------------------------------------------
%       FLUX  RXN ID    REACTION
----  ------  --------  ----------------------------------
100%    4.08  FALDA     fald_c + thf_c --> h2o_c + mlthf_c


In [133]:
AM1_OB3b_combined.metabolites.q8h2_im.summary()

PRODUCING REACTIONS -- Ubiquinol-8 (q8h2_im)
--------------------------------------------
%       FLUX  RXN ID    REACTION
----  ------  --------  --------------------------------------------------
100%      25  NADH16pp  4.0 h_c + nadh_c + q8_im --> 3.0 h_p + nad_c + ...

CONSUMING REACTIONS -- Ubiquinol-8 (q8h2_im)
--------------------------------------------
%       FLUX  RXN ID    REACTION
----  ------  --------  --------------------------------------------------
100%      25  PMMOipp   ch4_p + o2_p + q8h2_im --> h2o_p + meoh_p + q8_im


In [128]:
AM1_bigg.metabolites.qh2_c.summary()

PRODUCING REACTIONS -- UQH2 (qh2_c)
-----------------------------------
%       FLUX  RXN ID    REACTION
----  ------  --------  --------------------------------------------------
48%    1.95   SUCCD1    q_c + succ_c <=> fum_c + qh2_c
45%    1.83   NADH11    5.0 h_c + nadh_c + q_c --> 4.0 h_p + nad_c + qh2_c
7%     0.292  ETFDH     q_c + rETF_c <=> oxETF_c + qh2_c

CONSUMING REACTIONS -- UQH2 (qh2_c)
-----------------------------------
%       FLUX  RXN ID    REACTION
----  ------  --------  --------------------------------------------------
100%   4.08   PMMOipp   ch4_p + o2_p + qh2_c --> h2o_p + meoh_p + q_c


In [124]:
AM1_bigg.metabolites.focytc_c.summary()

PRODUCING REACTIONS -- reduced_cytochrome_c (focytc_c)
------------------------------------------------------
%       FLUX  RXN ID    REACTION
----  ------  --------  --------------------------------------------------
100%    8.16  MEOHDH    2.0 ficytc_c + meoh_p --> fald_p + 2.0 focytc_c...

CONSUMING REACTIONS -- reduced_cytochrome_c (focytc_c)
------------------------------------------------------
%       FLUX  RXN ID    REACTION
----  ------  --------  --------------------------------------------------
100%    8.16  CYO_2p    2.0 focytc_c + 4.0 h_c + 0.5 o2_c --> 2.0 ficyt...


In [384]:
AM1_bigg.metabolites.mlthf_c.summary()

PRODUCING REACTIONS -- 5_10_methylenetetrahydrofolate (mlthf_c)
---------------------------------------------------------------
%       FLUX  RXN ID    REACTION
----  ------  --------  -------------------------------------------
100%    4.08  FALDA     fald_c + thf_c --> h2o_c + mlthf_c

CONSUMING REACTIONS -- 5_10_methylenetetrahydrofolate (mlthf_c)
---------------------------------------------------------------
%       FLUX  RXN ID    REACTION
----  ------  --------  -------------------------------------------
64%     2.6   GHMT2r    gly_c + h2o_c + mlthf_c <=> ser_L_c + thf_c
36%     1.47  MTHFD     methf_c + nadph_c <=> mlthf_c + nadp_c


Check GEM reactions that have SP0001 gene reaction rule:

In [382]:
for reaction in AM1_bigg.reactions:
    if "SP" in reaction.gene_reaction_rule:
        print(reaction.id, reaction.name, reaction.reaction)

SPOTN 1_amino_propan_2_one_3_phosphate_synthesis ao4pob_c + h_c --> 3a2op_c + co2_c
FALDA H4F_dependent_formaldehyde_activating fald_c + thf_c --> h2o_c + mlthf_c
G5SADr L_glutamate_gamma_semialdehyde__cyclase_dehydratase_ glu5sa_c <=> 1pyr5c_c + h2o_c + h_c
OMCDC 2_isopropyl_3_oxosuccinate_decarboxylase 3c4mop_c + h_c --> 4mop_c + co2_c


Let's try uphill electron transfer as implemented in Methylococcus capsulatus Bath:

In [187]:
with AM1_bigg as model:
    reaction_id = "Biomass_Mextorquens_AM1_core"
    model.objective = reaction_id
    model.reactions.CYOD.knock_out()
    model.reactions.CYOO_4p.knock_out()
    AM1_bigg.reactions.PMMODCipp.knock_out()
    AM1_bigg.reactions.PMMOipp.bounds = (0, 1000)
    AM1_bigg.reactions.UQCYOR_2p.bounds = (-1000, 1000)
    model.optimize()
    model_objective = model.optimize().fluxes
    model_objective.to_csv(Escher_gapfilling_dir + reaction_id + "_UQCYOR_2p"+ "_AM1" + '.csv')
    model.summary()

IN FLUXES            OUT FLUXES    OBJECTIVES
-------------------  ------------  -----------------------
o2_e       5.46      h2o_e  7      Biomass_Mext...  0.0656
ch4_e      4.08      co2_e  1.44
nh4_e      0.641     h_e    0.146
pi_e       0.0458
k_e        0.0121
so4_e      0.00414
mg2_e      0.000541
fe2_e      0.000526
ca2_e      0.000325
cl_e       0.000325
na1_e      0.000267
mobd_e     0.000231
cu2_e      0.000216
mn2_e      0.000216
zn2_e      0.000216
cobalt2_e  1.44e-05


#### Study why uphill electron transfer does not work properly! 
Block UQCYOR_2R reaction in Bath and AM1 and see what happens:

In [217]:
with model_Bath as model:
    reaction_id = "BIOMASS_Mcapsulatus"
    model.objective = reaction_id
    model.reactions.SMMOi.knock_out()
    model.reactions.PMMODCipp.knock_out()
    model.reactions.PMMOipp.bounds = (0, 1000)
    model.reactions.CYOR_q8ppi.knock_out()
    model.optimize()
    model_objective = model.optimize().fluxes
    model_objective.to_csv(Escher_gapfilling_dir + reaction_id + "_CYOR_q8ppi_blocked"+ "_AM1" + '.csv')
    model.summary()
    print(" ")
    model.metabolites.ficytcc553_p.summary()
    print(" ")
    model.metabolites.q8h2_im.summary()
    print(" ")
    model.metabolites.nadh_c.summary()
    print(" ")
    model.metabolites.atp_c.summary()

IN FLUXES         OUT FLUXES    OBJECTIVES
----------------  ------------  ----------------------
o2_e   27.8       h2o_e  33     BIOMASS_Mcap...  0.178
ch4_e  18.5       co2_e  12.5
no3_e   1.46
h_e     1.46
pi_e    0.0981
cl_e    0.0382
k_e     0.0314
so4_e   0.03
mg2_e   0.022
ca2_e   0.0124
na1_e   0.00697
fe2_e   0.00109
cu2_e   0.000269
zn2_e   0.000178
 
PRODUCING REACTIONS -- ferricytochrome c (ficytcc...)
-----------------------------------------------------
%       FLUX  RXN ID      REACTION
----  ------  ----------  --------------------------------------------------
100%    36.9  CYTAA3ipp   4.0 focytcc553_p + 8.0 h_c + o2_c --> 4.0 ficyt...

CONSUMING REACTIONS -- ferricytochrome c (ficytcc...)
-----------------------------------------------------
%       FLUX  RXN ID      REACTION
----  ------  ----------  --------------------------------------------------
100%    36.9  CyLCyHORpp  ficytcc553_p + focytcc555_p <=> ficytcc555_p + ...
 
PRODUCING REACTIONS -- Ubiquinol-8 (q8h

In [216]:
with AM1_bigg as model:
    reaction_id = "Biomass_Mextorquens_AM1_core"
    model.objective = reaction_id
    model.reactions.CYOD.knock_out()
    model.reactions.CYOO_4p.knock_out()
    model.reactions.PMMODCipp.knock_out()
    model.reactions.PMMOipp.bounds = (0, 1000)
    model.reactions.UQCYOR_2p.knock_out()
    model.optimize()
    model_objective = model.optimize().fluxes
    model_objective.to_csv(Escher_gapfilling_dir + reaction_id + "_UQCYOR_2p_blocked"+ "_AM1" + '.csv')
    model.summary()
    print(" ")
    model.metabolites.ficytc_c.summary()
    print(" ")
    model.metabolites.qh2_c.summary()
    print(" ")
    model.metabolites.nadh_c.summary()
    print(" ")
    model.metabolites.atp_c.summary()

IN FLUXES            OUT FLUXES    OBJECTIVES
-------------------  ------------  -----------------------
o2_e       6.12      h2o_e  7.28   Biomass_Mext...  0.0495
ch4_e      4.08      co2_e  2.09
nh4_e      0.484     h_e    0.11
pi_e       0.0346
k_e        0.00913
so4_e      0.00312
mg2_e      0.000408
fe2_e      0.000397
ca2_e      0.000245
cl_e       0.000245
na1_e      0.000201
mobd_e     0.000174
cu2_e      0.000163
mn2_e      0.000163
zn2_e      0.000163
cobalt2_e  1.09e-05
 
PRODUCING REACTIONS -- oxidized_cytochrome_c (ficytc_c)
-------------------------------------------------------
%       FLUX  RXN ID    REACTION
----  ------  --------  --------------------------------------------------
100%    8.16  CYO_2p    2.0 focytc_c + 4.0 h_c + 0.5 o2_c --> 2.0 ficyt...

CONSUMING REACTIONS -- oxidized_cytochrome_c (ficytc_c)
-------------------------------------------------------
%       FLUX  RXN ID    REACTION
----  ------  --------  -----------------------------------------------

In [199]:
AM1_bigg.metabolites.ficytc_c.summary()

PRODUCING REACTIONS -- oxidized_cytochrome_c (ficytc_c)
-------------------------------------------------------
%       FLUX  RXN ID     REACTION
----  ------  ---------  --------------------------------------------------
68%     5.51  CYO_2p     2.0 focytc_c + 4.0 h_c + 0.5 o2_c --> 2.0 ficyt...
32%     2.65  UQCYOR_2p  2.0 ficytc_c + qh2_c <=> 2.0 focytc_c + 2.0 h_p...

CONSUMING REACTIONS -- oxidized_cytochrome_c (ficytc_c)
-------------------------------------------------------
%       FLUX  RXN ID     REACTION
----  ------  ---------  --------------------------------------------------
100%    8.16  MEOHDH     2.0 ficytc_c + meoh_p --> fald_p + 2.0 focytc_c...


In [207]:
AM1_bigg.reactions.MSUCCCOADH

<Reaction MSUCCCOADH at 0x7fd7de1e1748>

In [209]:
AM1_bigg.reactions.ETFDH

<Reaction ETFDH at 0x7fd7de870780>

## Central carbon metabolism addition/curation:
In OB3b, formate is the key entry point to the serine cycle and subsequent metabolism, not the formaldehyde. This is the main difference between OB3b and AM1
<br>Below is the exploration of AM1-style formaldehyde entry to serine cycle:

In [106]:
AM1_bigg.metabolites.methf_c.summary()

PRODUCING REACTIONS -- 5_10_methenyltetrahydrofolate (methf_c)
--------------------------------------------------------------
%       FLUX  RXN ID    REACTION
----  ------  --------  --------------------------------------
100%    1.47  MTHFD     methf_c + nadph_c <=> mlthf_c + nadp_c

CONSUMING REACTIONS -- 5_10_methenyltetrahydrofolate (methf_c)
--------------------------------------------------------------
%       FLUX  RXN ID    REACTION
----  ------  --------  --------------------------------------
100%    1.47  MTHFC     10fthf_c + h_c <=> h2o_c + methf_c


In [107]:
AM1_bigg.metabolites.get_by_id("10fthf_c").summary()

PRODUCING REACTIONS -- 10_formyltetrahydrofolate (10fthf_c)
-----------------------------------------------------------
%       FLUX  RXN ID    REACTION
----  ------  --------  -------------------------------------------------
100%  1.47    MTHFC     10fthf_c + h_c <=> h2o_c + methf_c

CONSUMING REACTIONS -- 10_formyltetrahydrofolate (10fthf_c)
-----------------------------------------------------------
%       FLUX  RXN ID    REACTION
----  ------  --------  -------------------------------------------------
98%   1.44    FTHFL     atp_c + for_c + thf_c <=> 10fthf_c + adp_c + pi_c
1%    0.0167  AICART    10fthf_c + aicar_c <=> fprica_c + thf_c


In [108]:
AM1_bigg.metabolites.get_by_id("for_c").summary()

PRODUCING REACTIONS -- formate (for_c)
--------------------------------------
%       FLUX  RXN ID    REACTION
----  ------  --------  -------------------------------------------------
100%    1.44  FTHFL     atp_c + for_c + thf_c <=> 10fthf_c + adp_c + pi_c

CONSUMING REACTIONS -- formate (for_c)
--------------------------------------
%       FLUX  RXN ID    REACTION
----  ------  --------  -------------------------------------------------
100%    1.44  FDHr      for_c + nad_c <=> co2_c + nadh_c


### Escher map drawing fixing of central carbon metabolism of OB3b 
Important reactions for central carbon metabolism include:
- H4MTP pathway
- H4Folate pathway
- Electron transport chain
- EMC pahtway 
- Serine cycle
- Some downstream pathways:
- Glycolysis and PPP pathway
<br>**Strategy:** 1. For each of the pathway try to refer to professional literature about the nature of this pathways
<br> 2. Use scripts from http://localhost:8888/notebooks/Scripts/GEM_OB3b/3.%20Pathway-to-pathway%20manual%20curation.ipynb amd your Carveme GEM for assisting in reconstructing OB3b central carbon metabolism pathways.

#### ETC

In [164]:
import re
for reaction in AM1_bigg.reactions:
    if re.search("ATP", reaction.name, re.IGNORECASE):
        print(reaction.id, reaction.name, reaction.reaction)

ATPM ATP_maintenance_requirement atp_c + h2o_c --> adp_c + h_c + pi_c
CU2abcpp copper_exporting_ATPase atp_c + cu2_c + h2o_c --> adp_c + cu2_p + h_c + pi_c
Kabcpp potassium_transporting_ATPase atp_c + h2o_c + k_p --> adp_c + h_c + k_c + pi_c
PRATPP phosphoribosyl_ATP_pyrophosphatase h2o_c + prbatp_c --> h_c + ppi_c + prbamp_c
ATPS4rpp ATP_synthase adp_c + 4.0 h_p + pi_c --> atp_c + h2o_c + 3.0 h_c
ATPPRTr ATP_phosphoribosyltransferase atp_c + prpp_c <=> ppi_c + prbatp_c
NTPP5 deoxynucleoside_triphosphate_pyrophosphohydrolase__dATP_ datp_c + h2o_c --> damp_c + h_c + ppi_c
NTPP6 nucleoside_triphosphate_pyrophosphohydrolase__ATP_ atp_c + h2o_c --> amp_c + h_c + ppi_c


In [165]:
import re
for reaction in AM1_bigg.reactions:
    if re.search("cytochrome", reaction.name, re.IGNORECASE):
        print(reaction.id, reaction.name, reaction.reaction)

CYO_2p cytochrome_c_oxidase 2.0 focytc_c + 4.0 h_c + 0.5 o2_c --> 2.0 ficytc_c + h2o_c + 2.0 h_p
CYOO_4p cytochrome_o_ubiquinol_oxidase 4.0 h_c + 0.5 o2_c + qh2_c --> h2o_c + 4.0 h_p + q_c
LDH_D3 D_lactate_dehydrogenase__cytochrome_ 2.0 ficytc_c + lac_D_c --> 2.0 focytc_c + 2.0 h_c + pyr_c
UQCYOR_2p ubiquinol_cytochrome_c_reductase 2.0 ficytc_c + qh2_c --> 2.0 focytc_c + 2.0 h_p + q_c
CYOD cytochrome_d_terminal_oxidase 4.0 h_c + o2_c + 2.0 qh2_c --> 2.0 h2o_c + 4.0 h_p + 2.0 q_c


In [166]:
import re
for reaction in AM1_bigg.reactions:
    if re.search("ubiquinone", reaction.name, re.IGNORECASE):
        print(reaction.id, reaction.name, reaction.reaction)

DMQMT2 3_demethylubiquinone_8_3_O_methyltransferase amet_c + dmuqh2_c --> ahcys_c + h_c + q8h2_c
NADH11 NADH_ubiquinone_oxidoreductase_I 5.0 h_c + nadh_c + q_c --> 4.0 h_p + nad_c + qh2_c


In [167]:
import re
for reaction in AM1_bigg.reactions:
    if re.search("transhydrogenase", reaction.name, re.IGNORECASE):
        print(reaction.id, reaction.name, reaction.reaction)

THD2pp NAD_P__transhydrogenase h_p + nadh_c + nadp_c --> h_c + nad_c + nadph_c


Here is the list for oxidative phosphorylation subsystem as implemented in AM1:

In [ ]:
ATPS4rpp --> ATP production
CYO_2p --> Complex IV, cytochrome c oxidase
CYOO_4p --> Complex IV
UQCYOR_2p --> type III complex, cytochrome c reductase
CYOD --> Complex IV, terminal oxidase, difference in cofactor used with Bath (qh2_c in AM1, ficytcc553_p in Bath)
NADH11 --> energy balance
THD2pp --> 

#### Formaldehyde oxidation
Use highly-curated reactions from carveme GEM and put them inside your AM1-based GEM

In [220]:
memote_directory = '/home/ensakz/Desktop/memote_m_trichosporium/'
draft_gems_dir_8 = 'draft_gems/8.OB3b_draft_gem_during_pathway_manual_curation/'
carveme_GEM = cobra.io.read_sbml_model(memote_directory + draft_gems_dir_8 + 'iMsOB3b_11_14_2019.xml')

In [310]:
bigg_OB3b = cobra.io.load_json_model(memote_directory + "AM1_bigg.json")

#### Add H4MTP pathway for potential formaldehyde oxidation:

In [242]:
H4MTP_pathway = ['FAEi', 'MTDBi', 'MCHi', 'FTR', 'FMFRDHi', 'FDH', 'FALDDHipp', 'ALDD1']
#other_pathways = ['ALDD1', 'FALDDHipp']
for reaction_id in H4MTP_pathway:
    try:
        bigg_OB3b.add_reaction(carveme_GEM.reactions.get_by_id(reaction_id))
    except:
        print(reaction_id, "is absent")
        
print("\nChecking whether H4MTP pathway reactions have been added to bigg_OB3b GEM:")
        
for reaction_id in H4MTP_pathway:
    try:
        print(bigg_OB3b.reactions.get_by_id(reaction_id).id)
    except:
        print(reaction_id, "is absent")

Ignoring reaction 'FAEi' since it already exists.
Ignoring reaction 'MTDBi' since it already exists.
Ignoring reaction 'MCHi' since it already exists.
Ignoring reaction 'FTR' since it already exists.
Ignoring reaction 'FMFRDHi' since it already exists.
Ignoring reaction 'FDH' since it already exists.
Ignoring reaction 'ALDD1' since it already exists.


FALDDHipp is absent

Checking whether H4MTP pathway reactions have been added to bigg_OB3b GEM:
FAEi
MTDBi
MCHi
FTR
FMFRDHi
FDH
FALDDHipp is absent
ALDD1


H4MTP pathway added:

In [247]:
draft_gems_dir_8 = 'draft_gems/8.OB3b_draft_gem_during_pathway_manual_curation/'
cobra.io.save_json_model(bigg_OB3b, memote_directory + draft_gems_dir_8 + "bigg_OB3b_25_11_2019.json")

Let's try to see how flux goes in simulation after H4MTP-pathway addition:

In [311]:
model_version = "bigg_OB3b_25_11_2019/"
with bigg_OB3b as model:
    reaction_id = "Biomass_Mextorquens_AM1_core"
    model.objective = reaction_id
    model.optimize()
    model_objective = model.optimize().fluxes
    model_objective.to_csv(Escher_gapfilling_dir + model_version + reaction_id + "_PMMOipp"+ "_AM1" + '.csv')
    model.summary()

IN FLUXES            OUT FLUXES    OBJECTIVES
-------------------  ------------  -----------------------
o2_e       6.12      h2o_e  7.28   Biomass_Mext...  0.0495
ch4_e      4.08      co2_e  2.09
nh4_e      0.484     h_e    0.11
pi_e       0.0346
k_e        0.00913
so4_e      0.00312
mg2_e      0.000408
fe2_e      0.000397
ca2_e      0.000245
cl_e       0.000245
na1_e      0.000201
mobd_e     0.000174
cu2_e      0.000163
mn2_e      0.000163
zn2_e      0.000163
cobalt2_e  1.09e-05


Let's check where mlthf_c (5_10_methylenetetrahydrofolate) comes from in bigg_OB3b_25_11_2019.json GEM: 

In [312]:
for reaction in bigg_OB3b.metabolites.mlthf_c.reactions:
    print(reaction.id, reaction.name, reaction.gene_reaction_rule, reaction)

GHMT2r serine_hydroxymethyltransferase CQW49_RS06375 GHMT2r: gly_c + h2o_c + mlthf_c <=> ser_L_c + thf_c
MTHFD methylenetetrahydrofolate_dehydrogenase CQW49_RS12245 MTHFD: methf_c + nadph_c <=> mlthf_c + nadp_c
MOHMT 3_methyl_2_oxobutanoate_hydroxymethyltransferase CQW49_RS13560 MOHMT: 3mob_c + h2o_c + mlthf_c <=> 2dhp_c + thf_c
FALDA H4F_dependent_formaldehyde_activating SP0001 FALDA: fald_c + thf_c --> h2o_c + mlthf_c
MTHFR3 5_10_methylenetetrahydrofolate_reductase CQW49_RS19120 MTHFR3: 2.0 h_c + mlthf_c + nadph_c <=> 5mthf_c + nadp_c
TMDS thymidylate_synthetase META1_2853 TMDS: dump_c + mlthf_c <=> dhf_c + dtmp_c
GLYCL_3 H_protein_aminomethyltransferase CQW49_RS13400 or CQW49_RS20565 GLYCL_3: h_c + pSamdhll_c + thf_c --> Gcvpdhll_c + mlthf_c + nh4_c


In [313]:
bigg_OB3b.metabolites.mlthf_c.summary()

PRODUCING REACTIONS -- 5_10_methylenetetrahydrofolate (mlthf_c)
---------------------------------------------------------------
%       FLUX  RXN ID    REACTION
----  ------  --------  -------------------------------------------
100%    3.81  FALDA     fald_c + thf_c --> h2o_c + mlthf_c

CONSUMING REACTIONS -- 5_10_methylenetetrahydrofolate (mlthf_c)
---------------------------------------------------------------
%       FLUX  RXN ID    REACTION
----  ------  --------  -------------------------------------------
67%     2.56  MTHFD     methf_c + nadph_c <=> mlthf_c + nadp_c
33%     1.24  GHMT2r    gly_c + h2o_c + mlthf_c <=> ser_L_c + thf_c


FALDA is the spontaneous (non-enzymatic) condensation of formaldehyde with H4F. Let's see what happens with FALDA knocked out:

In [314]:
with bigg_OB3b as model:
    biomass_id = "Biomass_Mextorquens_AM1_core"
    model.objective = biomass_id
    reaction_id = "FALDA_ko"
    model.reactions.FALDA.knock_out()
    model.optimize()
    model_objective = model.optimize().fluxes
    model_objective.to_csv(Escher_gapfilling_dir + model_version + reaction_id + "_PMMOipp"+ "_AM1" + '.csv')
    model.summary()
    bigg_OB3b.metabolites.fald_c.summary()
    bigg_OB3b.metabolites.mlthf_c.summary()

IN FLUXES            OUT FLUXES    OBJECTIVES
-------------------  ------------  -----------------------
o2_e       6.12      h2o_e  7.28   Biomass_Mext...  0.0495
ch4_e      4.08      co2_e  2.09
nh4_e      0.484     h_e    0.11
pi_e       0.0346
k_e        0.00913
so4_e      0.00312
mg2_e      0.000408
fe2_e      0.000397
cl_e       0.000245
ca2_e      0.000245
na1_e      0.000201
mobd_e     0.000174
cu2_e      0.000163
mn2_e      0.000163
zn2_e      0.000163
cobalt2_e  1.09e-05
PRODUCING REACTIONS -- formaldehyde (fald_c)
--------------------------------------------
%       FLUX  RXN ID    REACTION
----  ------  --------  --------------------------------------
100%    4.08  FALDtpp   fald_p <=> fald_c

CONSUMING REACTIONS -- formaldehyde (fald_c)
--------------------------------------------
%       FLUX  RXN ID    REACTION
----  ------  --------  --------------------------------------
100%    4.08  MPTFALDA  fald_c + h4mpt_c --> h2o_c + mlh4mpt_c
PRODUCING REACTIONS -- 5_10_methylen

H4MTP pathway is already present in the GEM, there is no reason to add alternative reactions to this GEM. --> delete H4MTP pathway alternatives from OB3b or load clean bigg_OB3b GEM again:

In [263]:
bigg_OB3b = cobra.io.load_json_model(memote_directory + "AM1_bigg.json")

In [264]:
import re
def reaction_id_search(model_from_list, word_to_search):
    for reaction in model_from_list.reactions:
        if re.search(word_to_search, reaction.name, re.IGNORECASE):
            print(reaction.id, reaction.name, reaction.reaction, reaction.gene_reaction_rule)

In [266]:
reaction_id_search(bigg_OB3b, 'formaldehyde')

FALDHp formaldehyde_dehydrogenase fald_p + 2.0 ficytc_c + h2o_p --> 2.0 focytc_c + for_p + 3.0 h_p CQW49_RS14430
FALDA H4F_dependent_formaldehyde_activating fald_c + thf_c --> h2o_c + mlthf_c SP0001
EX_fald_e exchange_flux___formaldehyde fald_e -->  
MPTFALDA H4MPT_dependent_formaldehyde_activating fald_c + h4mpt_c --> h2o_c + mlh4mpt_c CQW49_RS13515
FALDtpp formaldehyde_transport_via_diffusion__periplasm_to_cytoplasm_ fald_p <=> fald_c 
FALDtex formaldehyde_transport_via_diffusion__extracellular_to_periplasm_ fald_e <=> fald_p 


In [268]:
reaction_id_search(bigg_OB3b, 'formaldehyde')

FALDHp formaldehyde_dehydrogenase fald_p + 2.0 ficytc_c + h2o_p --> 2.0 focytc_c + for_p + 3.0 h_p CQW49_RS14430
FALDA H4F_dependent_formaldehyde_activating fald_c + thf_c --> h2o_c + mlthf_c SP0001
EX_fald_e exchange_flux___formaldehyde fald_e -->  
MPTFALDA H4MPT_dependent_formaldehyde_activating fald_c + h4mpt_c --> h2o_c + mlh4mpt_c CQW49_RS13515
FALDtpp formaldehyde_transport_via_diffusion__periplasm_to_cytoplasm_ fald_p <=> fald_c 
FALDtex formaldehyde_transport_via_diffusion__extracellular_to_periplasm_ fald_e <=> fald_p 


In [317]:
bigg_OB3b.add_reaction(carveme_GEM.reactions.ALDD1)

Ignoring reaction 'ALDD1' since it already exists.


**Conclusion: FALDA is spontaneous reaction, and there is evidence that formaldehyde assimilation into serine cycle is done via H4MTP and downstream pathways. To keep flux going through, H4MTP pathway, I have to knockout FALDA**
This is interesting as inclusion of ALDD1 reaction makes all flux go through it for H4MTP pathway.

Set reaction constraint of FALDHp to (0, 0) as there almost no evidence this reaction is present/active in OB3b. (Based on ortholog analysis and inspection of other models (look at section 4 Formaldehyde and formate oxidation) + literature review, there is no evidence of the presence of this enzyme in OB3b.)

In [329]:
bigg_OB3b.reactions.FALDHp.bounds = (0, 0)

#### Formate to CO2
NADH-dependent formate dehydrogenase

#### Check serine cycle

In [332]:
bigg_OB3b.metabolites.oaa_c.summary()

PRODUCING REACTIONS -- oxaloacetate (oaa_c)
-------------------------------------------
%       FLUX  RXN ID    REACTION
----  ------  --------  -----------------------------------------------
100%  0.561   PPC       co2_c + h2o_c + pep_c --> h_c + oaa_c + pi_c

CONSUMING REACTIONS -- oxaloacetate (oaa_c)
-------------------------------------------
%       FLUX  RXN ID    REACTION
----  ------  --------  -----------------------------------------------
70%   0.393   MDH       mal_L_c + nad_c <=> h_c + nadh_c + oaa_c
20%   0.111   ASPTA     akg_c + asp_L_c <=> glu_L_c + oaa_c
10%   0.0569  CS        accoa_c + h2o_c + oaa_c <=> cit_c + coa_c + h_c


In [334]:
bigg_OB3b.reactions.ASPTA

<Reaction ASPTA at 0x7fd7d85fff98>

In [333]:
bigg_OB3b.reactions.CS

<Reaction CS at 0x7fd7d8249a90>

ACITL reaction from Escher map which is ATP-dependent citrate synthase is absent in OB3b GEM:

In [335]:
bigg_OB3b.reactions.ACITL

AttributeError: DictList has no attribute or entry ACITL

**Note:** 2_oxoglutarate_dehydrogenase has no flux in OB3b GEM

In [338]:
bigg_OB3b.metabolites.akg_c.summary()

PRODUCING REACTIONS -- 2_oxoglutarate (akg_c)
---------------------------------------------
%        FLUX  RXN ID    REACTION
----  -------  --------  --------------------------------------------------
28%   0.111    ASPTA     akg_c + asp_L_c <=> glu_L_c + oaa_c
23%   0.0894   PSERTr    3php_c + glu_L_c <=> akg_c + pser_L_c
14%   0.0569   ICDHyr    icit_c + nadp_c --> akg_c + co2_c + nadph_c
14%   0.0546   ALATA_D   glu_D_c + pyr_c <=> akg_c + ala_D_c
5%    0.0208   VALTA     3mob_c + glu_L_c <=> akg_c + val_L_c
5%    0.0196   LEUTAi    4mop_c + glu_L_c <=> akg_c + leu_L_c
4%    0.0142   ACOTA     acorn_c + akg_c <=> acg5sa_c + glu_L_c
3%    0.0128   SDPTA     glu_L_c + sl2a6o_c <=> akg_c + sl26da_c
3%    0.0103   ILETA     3mop_c + glu_L_c <=> akg_c + ile_L_c
2%    0.00673  PHETA1    glu_L_c + phpyr_c <=> akg_c + phe_L_c

CONSUMING REACTIONS -- 2_oxoglutarate (akg_c)
---------------------------------------------
%        FLUX  RXN ID    REACTION
----  -------  --------  --------------

In [339]:
bigg_OB3b.reactions.AKGDH

<Reaction AKGDH at 0x7fd7d8712ac8>

In Bath, corresponding reactions are knocked out (2_oxoglutarate_dehydrogenase).

In [341]:
model_Bath.reactions.AKGDH.bounds
model_Bath.reactions.OOR3r

<Reaction OOR3r at 0x7fd7dccf6da0>

In [342]:
bigg_OB3b.metabolites.succ_c.summary()

PRODUCING REACTIONS -- succinate (succ_c)
-----------------------------------------
%       FLUX  RXN ID    REACTION
----  ------  --------  --------------------------------------------------
96%   0.279   SUCOAS    atp_c + coa_c + succ_c <=> adp_c + pi_c + succoa_c
4%    0.0128  SDPDS     h2o_c + sl26da_c <=> 26dap_LL_c + succ_c

CONSUMING REACTIONS -- succinate (succ_c)
-----------------------------------------
%       FLUX  RXN ID    REACTION
----  ------  --------  --------------------------------------------------
100%  0.292   SUCCD1    q_c + succ_c <=> fum_c + qh2_c


In [343]:
bigg_OB3b.reactions.FRD

AttributeError: DictList has no attribute or entry FRD

In [345]:
bigg_OB3b.metabolites.fum_c.summary()

PRODUCING REACTIONS -- fumarate (fum_c)
---------------------------------------
%       FLUX  RXN ID    REACTION
----  ------  --------  ------------------------------
89%   0.292   SUCCD1    q_c + succ_c <=> fum_c + qh2_c
4%    0.0134  ADSL2r    25aics_c <=> aicar_c + fum_c
4%    0.0119  ARGSL     argsuc_c <=> arg_L_c + fum_c
3%    0.0107  ADSL1r    dcamp_c <=> amp_c + fum_c

CONSUMING REACTIONS -- fumarate (fum_c)
---------------------------------------
%       FLUX  RXN ID    REACTION
----  ------  --------  ------------------------------
100%  0.328   FUM       fum_c + h2o_c <=> mal_L_c


#### EMC pathway

In [348]:
bigg_OB3b.metabolites.r3hbcoa_c.summary()

PRODUCING REACTIONS -- R__3_hydroxybutanoyl_CoA (r3hbcoa_c)
-----------------------------------------------------------
%       FLUX  RXN ID    REACTION
----  ------  --------  ----------------------------------------------
100%  0.0137  AACOAR    aacoa_c + h_c + nadph_c --> nadp_c + r3hbcoa_c

CONSUMING REACTIONS -- R__3_hydroxybutanoyl_CoA (r3hbcoa_c)
-----------------------------------------------------------
%       FLUX  RXN ID    REACTION
----  ------  --------  ----------------------------------------------
100%  0.0137  PHAS      r3hbcoa_c --> coa_c + pbhb_c


In [349]:
bigg_OB3b.reactions.AACOAR

<Reaction AACOAR at 0x7fd7d831d320>

**Note:** Pathway interconversions and cofactors are important. Think about them

## Check GEM for sanity:
One way to check GEM for sanity is to perform memote snapshot report.
Another is to do several functional tests based on other GEM papers:
Examples of such functional tests are given in Methylococcus capsulatus Bath GEM complimentary scripts and Methylocystis GEM paper.

Reference used to check GEM sanity:
- https://microbialcellfactories.biomedcentral.com/articles/10.1186/s12934-019-1154-5#Sec9 +
- https://www.sciencedirect.com/science/article/pii/S1096717618304592?via%3Dihub +
- https://bmcsystbiol.biomedcentral.com/articles/10.1186/1752-0509-5-189#Sec18 +
- 
- https://memote.io/
- Personal OneNote notes +
- Previous Jupyter notebook scripts +

### Step 1: Generate memote snapshot report
Report was generated using https://memote.io/ --> takes too much time. Finally, report was generated using memote local version:

In [ ]:
$ memote report snapshot path/to/model.xml

Inspect report:

Memote snapshot report looks really good! It can be inspected in greater detail later

### Step 2: Methylococcus capsulatus GEM complimentary scripts
Make a list of key metabolites sanity of which should be checked:
1. ATP: Reaction directionality was modified in order to avoid unfeasible ATP production, so that the model only simulates ATP production in the respiratory chain and in the lower glycolysis. +
2. H+: Directionality of the reactions involving proton transport was also modified in order to guarantee that no proton flux towards the exterior of the cell is allowed if it is not mediated by the components of the respiratory chain or associated to some energy cost. This was done in order to avoid proton flux against its concentration gradient. +

#### ATP

In [361]:
for reaction in bigg_OB3b.metabolites.atp_c.reactions:
    if "Biomass" in reaction.id:
        pass
    else:
        print(reaction)

ARGTRS: arg_L_c + atp_c + trnaarg_c --> amp_c + argtrna_c + ppi_c
FMNAT: atp_c + fmn_c + h_c --> fad_c + ppi_c
ACKr: ac_c + atp_c <=> actp_c + adp_c
PMPK: 4ampm_c + atp_c <=> 2mahmp_c + adp_c
SUCOAS: atp_c + coa_c + succ_c <=> adp_c + pi_c + succoa_c
FACOAL160: atp_c + coa_c + hdca_c <=> amp_c + pmtcoa_c + ppi_c
ASPK: asp_L_c + atp_c <=> 4pasp_c + adp_c
CBLAT: atp_c + cbl1_c + h_c --> adocbl_c + pppi_c
GMPS: atp_c + nh4_c + xmp_c --> amp_c + gmp_c + 2.0 h_c + ppi_c
FTHFL: atp_c + for_c + thf_c <=> 10fthf_c + adp_c + pi_c
ADOCBIK: adocbi_c + atp_c <=> adocbip_c + adp_c + h_c
HMPK1: 4ahmmp_c + atp_c <=> 4ampm_c + adp_c + h_c
RBFK: atp_c + ribflv_c --> adp_c + fmn_c + h_c
PRBK: atp_c + ru5p_D_c <=> adp_c + h_c + rb15bp_c
CBIAT: atp_c + cbi_c + h_c --> adocbi_c + pppi_c
ATPPRTr: atp_c + prpp_c <=> ppi_c + prbatp_c
PRAGSr: atp_c + gly_c + pram_c <=> adp_c + gar_c + h_c + pi_c
ILETRS: atp_c + ile_L_c + trnaile_c --> amp_c + iletrna_c + ppi_c
FACOAL100: atp_c + coa_c + dca_c <=> amp_c + dcaco

All reactions that can produce ATP:

In [363]:
model=bigg_OB3b
[rxn for rxn in model.metabolites.atp_c.reactions if (model.metabolites.atp_c in rxn.products or rxn.reversibility) and abs(rxn.flux) > 0]

[<Reaction ACKr at 0x7fd7d83b4048>,
 <Reaction PMPK at 0x7fd7d85160b8>,
 <Reaction SUCOAS at 0x7fd7d86520f0>,
 <Reaction FACOAL160 at 0x7fd7d85fa0f0>,
 <Reaction ASPK at 0x7fd7d8442128>,
 <Reaction FTHFL at 0x7fd7d85741d0>,
 <Reaction ATPPRTr at 0x7fd7d83b4320>,
 <Reaction PRAGSr at 0x7fd7d820e320>,
 <Reaction SKMK at 0x7fd7d83924e0>,
 <Reaction BIOTINS at 0x7fd7d82e0518>,
 <Reaction GK1 at 0x7fd7d82206d8>,
 <Reaction PYK at 0x7fd7d85c6898>,
 <Reaction FACOAL180 at 0x7fd7d8574940>,
 <Reaction UMPK at 0x7fd7d8418a58>,
 <Reaction PRASCS at 0x7fd7d8418b70>,
 <Reaction ALAALAr at 0x7fd7d8284cf8>,
 <Reaction UAMAS at 0x7fd7d878cdd8>,
 <Reaction ADSK at 0x7fd7d869ae48>,
 <Reaction PGK at 0x7fd7d84cce48>,
 <Reaction ADCYRS at 0x7fd7d842af28>,
 <Reaction SADT at 0x7fd7d8382f60>,
 <Reaction HSK at 0x7fd7d859d1d0>,
 <Reaction ADK1 at 0x7fd7d876b2e8>,
 <Reaction CBPS at 0x7fd7d859d3c8>,
 <Reaction GLUCYS at 0x7fd7d8407470>,
 <Reaction DTMPK at 0x7fd7d8783588>,
 <Reaction ATPS4rpp at 0x7fd7d845371

Reactions that where atp is a product, reaction is reversible and atp is produced:

In [390]:
[rxn for rxn in model.metabolites.atp_c.reactions if (model.metabolites.atp_c in rxn.products and rxn.reversibility) and abs(rxn.flux) > 0]

[<Reaction BIOTINS at 0x7fd7d82e0518>]

BIOTINS is a gapfilling reaction:

In [405]:
reaction = bigg_OB3b.reactions.BIOTINS
with bigg_OB3b as model:
    reaction.bounds = (-1000, 1000)
    print(model.optimize())
reaction.bounds = (-1000, 1000)
reaction.gene_reaction_rule = "SP0001"

<Solution 0.050 at 0x7fd7d8f6ec50>


Reactions where atp is a reactant, reaction is reversible and  atp is produced:

In [408]:
atp_in_reactants=[rxn for rxn in model.metabolites.atp_c.reactions if (model.metabolites.atp_c in rxn.reactants and rxn.reversibility) and abs(rxn.flux) > 0]

In [418]:
with bigg_OB3b as model:
    for reaction in atp_in_reactants:
        if reaction.flux > 0:
            print(reaction.id, reaction.flux)
            bigg_OB3b.reactions.get_by_id(reaction.id).bounds = (0, 1000)
        else:
            pass
            #print(reaction.id)
    print(model.optimize())
    
for reaction in atp_in_reactants:
    if reaction.flux > 0:
        bigg_OB3b.reactions.get_by_id(reaction.id).bounds = (0, 1000)
print(model.optimize())

ASPK 0.04112718403702419
FTHFL 0.030074790479727834
ATPPRTr 0.0033543670203390772
PRAGSr 0.013365646989332096
SKMK 0.010522526529696676
GK1 0.006016042375791347
UMPK 0.008699336006188762
PRASCS 0.013354776470056584
ALAALAr 4.950145389577022e-06
UAMAS 4.950145389577022e-06
ADSK 0.002922651970536141
PGK 0.18253907448250312
ADCYRS 1.087051927551114e-05
SADT 0.002922651970536141
HSK 0.025564511049350012
ADK1 0.36067886654569525
CBPS 0.020619781118829186
GLUCYS 1.087051927551114e-05
DTMPK 0.0007567782271585335
ACGK 0.01417652337539405
PTPAT 1.0870519275600022e-05
GTHS 1.087051927551114e-05
TDSK 4.950145389577022e-06
HPPK2 4.0511989869381786e-06
CYTK1 0.003392376761214487
FACOAL161 4.448992368209088e-17
HGBAMS 1.087051927551114e-05
<Solution 0.050 at 0x7fd7daaa3c88>
<Solution 0.050 at 0x7fd7daaa3828>


In [427]:
[rxn for rxn in model.metabolites.atp_c.reactions if (model.metabolites.atp_c in rxn.products or rxn.reversibility) and abs(rxn.flux) > 0]

[<Reaction BIOTINS at 0x7fd7d82e0518>,
 <Reaction FACOAL180 at 0x7fd7d8574940>,
 <Reaction FACOAL181 at 0x7fd7d8574cc0>,
 <Reaction ATPS4rpp at 0x7fd7d8453710>]

In [426]:
reaction = bigg_OB3b.reactions.PYK
with bigg_OB3b as model:
    reaction.bounds = (0, 1000)
    print(model.optimize())
reaction.bounds = (0, 1000)

<Solution 0.050 at 0x7fd7d93a59e8>


In [433]:
reaction = bigg_OB3b.reactions.FACOAL180
print(reaction)
with bigg_OB3b as model:
    reaction.bounds = (0, 1000)
    print(model.optimize())
reaction.bounds = (0, 1000)

FACOAL180: atp_c + coa_c + ocdca_c --> amp_c + ppi_c + stcoa_c
<Solution 0.050 at 0x7fd7d93a5208>


In [432]:
reaction = bigg_OB3b.reactions.FACOAL181
print(reaction)
with bigg_OB3b as model:
    reaction.bounds = (0, 1000)
    print(model.optimize())
reaction.bounds = (0, 1000)

FACOAL181: atp_c + coa_c + ocdcea_c <=> amp_c + odecoa_c + ppi_c
<Solution 0.050 at 0x7fd7d9f8fdd8>


Finally, only BIOTINS (gapfilling reaction) and ATPS4rpp is allowed to produce ATP in the system:

In [466]:
[rxn for rxn in model.metabolites.atp_c.reactions if (model.metabolites.atp_c in rxn.products or rxn.reversibility) and abs(rxn.flux) > 0]

[<Reaction BIOTINS at 0x7fd7d82e0518>, <Reaction ATPS4rpp at 0x7fd7d8453710>]

Fixing ATP maintenance requirement:

In [464]:
bigg_OB3b.reactions.ATPM.gene_reaction_rule = "SP0001"

The value for growth associated maintenance was taken from:
https://microbialcellfactories.biomedcentral.com/articles/10.1186/s12934-019-1154-5
<br><br>In order to obtain an estimation of the non-growth associated ATP requirements (which in the absence of methane are supplied by endogenous respiration), oxygen consumption was monitored following total methane depletion (in triplicate serum bottles), leading to a value of 0.71 ± 0.2 mmolO2 g-DW−1 h−1. In this context, assuming five molecules of ATP produced per oxygen molecule, the non-growth associated ATP maintenance would be 3.5 mmol ATP g-DW−1 h−1.

In [461]:
bigg_OB3b.reactions.ATPM.bounds = (3.5,3.5)

In [462]:
bigg_OB3b.summary()

IN FLUXES            OUT FLUXES    OBJECTIVES
-------------------  ------------  -----------------------
o2_e       6.4       h2o_e  7.4    Biomass_Mext...  0.0429
ch4_e      4.08      co2_e  2.36
nh4_e      0.419
h_e        0.146
pi_e       0.03
k_e        0.00791
so4_e      0.00271
mg2_e      0.000354
fe2_e      0.000344
cl_e       0.000212
ca2_e      0.000212
na1_e      0.000175
mobd_e     0.000151
cu2_e      0.000142
mn2_e      0.000142
zn2_e      0.000142
cobalt2_e  9.42e-06


In [463]:
bigg_OB3b.metabolites.atp_c.summary()

PRODUCING REACTIONS -- ATP (atp_c)
----------------------------------
%       FLUX  RXN ID      REACTION
----  ------  ----------  --------------------------------------------------
100%   8.76   ATPS4rpp    adp_c + 4.0 h_p + pi_c --> atp_c + h2o_c + 3.0 h_c

CONSUMING REACTIONS -- ATP (atp_c)
----------------------------------
%       FLUX  RXN ID      REACTION
----  ------  ----------  --------------------------------------------------
40%    3.5    ATPM        atp_c + h2o_c --> adp_c + h_c + pi_c
26%    2.31   Biomass...  0.034671 14bglu_c + 0.033189 acgam_c + 0.000219...
10%    0.877  GLYCK2      atp_c + glyc_R_c --> 2pg_c + adp_c + h_c
7%     0.624  MALCOAS     atp_c + coa_c + mal_L_c --> adp_c + malScoa_c +...
4%     0.313  ADK1        amp_c + atp_c --> 2.0 adp_c
3%     0.253  PPCOAC      atp_c + hco3_c + ppcoa_c --> adp_c + h_c + mmco...
2%     0.158  PGK         3pg_c + atp_c --> 13dpg_c + adp_c


Check ATP-producing reactions again setting objective as ATP exchange reaction:

In [468]:
model_universal = cobra.io.load_json_model(models_directory + "universal_model.json")

In [473]:
model_universal.reactions.EX_atp_e.reaction = 'atp_c --> '
model_universal.reactions.EX_atp_e.id = 'EX_atp_c'

"The proof for this is that when we optimize the production of ATP and at the same time have an insanely high turnover of ATP when the model is carbon limited, the solution will be infeasible." --> Taken from 8. Corrections to Validation Methylococcus capsulatus complimentary scritps

In [490]:
with bigg_OB3b as model:
    model.add_reaction(model_universal.reactions.EX_atp_c)
    model.objective = "EX_atp_c"
    model.reactions.ATPS4rpp.bounds = (100, 100)
    model.optimize()
    for rxn in model.metabolites.atp_c.reactions:
        if rxn.flux > 1:
            print (rxn.id,rxn.flux)

SUCOAS 96.49999999999999
ATPM 3.5
ATPS4rpp 100.0


cobra/util/solver.py:416 UserWarning: solver status is 'infeasible'


Let's fix this:
<br>Reaction directionality was modified in order to avoid unfeasible ATP production, so that the model only simulates ATP production in the respiratory chain and in the lower glycolysis.
- Respiratory chain --> ATPase
- Lower glycolysis:
2 reactions that produce ATP in lower part of glycolysis:
1) phosphoglycerokinase
2) pyruvate kinase

--> https://laboratoryinfo.com/wp-content/uploads/2016/02/glycolysis.png

In [492]:
reaction_id_search(bigg_OB3b, "phosphogly")

PGK phosphoglycerate_kinase 3pg_c + atp_c --> 13dpg_c + adp_c CQW49_RS11910
PGM phosphoglyceromutase 2pg_c <=> 3pg_c CQW49_RS04790 or CQW49_RS21290
PGCDr D_3_phosphoglycerate_dehydrogenase 3pg_c + nad_c <=> 3php_c + h_c + nadh_c CQW49_RS10875
PGLYCP phosphoglycolate_phosphatase 2pglyc_c + h2o_c --> glyclt_c + pi_c CQW49_RS14255
DPGM bisphosphoglycerate_mutase 13dpg_c <=> 23dpg_c + h_c CQW49_RS04790 or CQW49_RS21290


In [494]:
bigg_OB3b.reactions.PGK.bounds = (-1000, 1000)

In [498]:
reaction_id_search(bigg_OB3b, "pyruvate")

EX_pyr_e exchange_flux___pyruvate pyr_e -->  
PYRtex pyruvate_transport_via_diffusion__extracellular_to_periplasm_ pyr_e <=> pyr_p 
PPDK pyruvate_phosphate_dikinase atp_c + pi_c + pyr_c --> amp_c + h_c + pep_c + ppi_c CQW49_RS04525
PYK pyruvate_kinase atp_c + pyr_c --> adp_c + h_c + pep_c CQW49_RS01150
MCPST 3_mercaptopyruvate_sulfurtransferase cyan_c + mercppyr_c --> h_c + pyr_c + tcynt_c CQW49_RS20755
PPC phosphoenolpyruvate_carboxylase co2_c + h2o_c + pep_c --> h_c + oaa_c + pi_c CQW49_RS12265
PYRt2rpp pyruvate_symport__periplasm_to_cytoplasm_ h_p + pyr_p --> h_c + pyr_c 
PDH pyruvate_dehydrogenase coa_c + nad_c + pyr_c --> accoa_c + co2_c + nadh_c CQW49_RS01945 and CQW49_RS01950 and CQW49_RS01955 and CQW49_RS06870
GLYD hydroxypyruvate_reductase h_c + hpyr_c + nadh_c <=> glyc_R_c + nad_c CQW49_RS12240


In [500]:
reaction = bigg_OB3b.reactions.PYK
print(reaction)
with bigg_OB3b as model:
    reaction.bounds = (-1000, 1000)
    print(model.optimize())
reaction.bounds = (-1000, 1000)

PYK: atp_c + pyr_c --> adp_c + h_c + pep_c
<Solution 0.043 at 0x7fd7d62a0d68>


See how ATP distribution and growth rate changes (growth rate up from 0.0429 to 0.0433):

In [504]:
bigg_OB3b.metabolites.atp_c.summary()

PRODUCING REACTIONS -- ATP (atp_c)
----------------------------------
%      FLUX  RXN ID      REACTION
---  ------  ----------  --------------------------------------------------
98%   8.68   ATPS4rpp    adp_c + 4.0 h_p + pi_c --> atp_c + h2o_c + 3.0 h_c
2%    0.139  PYK         atp_c + pyr_c <=> adp_c + h_c + pep_c

CONSUMING REACTIONS -- ATP (atp_c)
----------------------------------
%      FLUX  RXN ID      REACTION
---  ------  ----------  --------------------------------------------------
40%   3.5    ATPM        atp_c + h2o_c --> adp_c + h_c + pi_c
26%   2.33   Biomass...  0.034671 14bglu_c + 0.033189 acgam_c + 0.000219...
10%   0.887  GLYCK2      atp_c + glyc_R_c --> 2pg_c + adp_c + h_c
7%    0.631  MALCOAS     atp_c + coa_c + mal_L_c --> adp_c + malScoa_c +...
4%    0.316  ADK1        amp_c + atp_c --> 2.0 adp_c
3%    0.256  PPCOAC      atp_c + hco3_c + ppcoa_c --> adp_c + h_c + mmco...
2%    0.16   PGK         3pg_c + atp_c <=> 13dpg_c + adp_c


In [503]:
bigg_OB3b.summary()

IN FLUXES            OUT FLUXES    OBJECTIVES
-------------------  ------------  -----------------------
o2_e       6.38      h2o_e  7.39   Biomass_Mext...  0.0433
ch4_e      4.08      co2_e  2.34
nh4_e      0.424
h_e        0.147
pi_e       0.0303
k_e        0.008
so4_e      0.00274
mg2_e      0.000358
fe2_e      0.000348
ca2_e      0.000215
cl_e       0.000215
na1_e      0.000176
mobd_e     0.000153
cu2_e      0.000143
mn2_e      0.000143
zn2_e      0.000143
cobalt2_e  9.52e-06


In [511]:
[rxn for rxn in bigg_OB3b.metabolites.atp_c.reactions if (bigg_OB3b.metabolites.atp_c in rxn.products or rxn.reversibility) and abs(rxn.flux) > 0]

[<Reaction BIOTINS at 0x7fd7d82e0518>,
 <Reaction PYK at 0x7fd7d85c6898>,
 <Reaction PGK at 0x7fd7d84cce48>,
 <Reaction ATPS4rpp at 0x7fd7d8453710>]

Save ATP-consistent GEM:

In [510]:
cobra.io.save_json_model(bigg_OB3b, memote_directory + draft_gems_dir_8 + "bigg_OB3b_26_11_2019.json")

#### H+ proton leakage outside the GEM
Let's fix this: H+ --> Directionality of the reactions involving proton transport was also modified in order to guarantee that no proton flux towards the exterior of the cell is allowed if it is not mediated by the components of the respiratory chain or associated to some energy cost. This was done in order to avoid proton flux against its concentration gradient. 

In [514]:
bigg_OB3b.metabolites.h_p.summary()

PRODUCING REACTIONS -- H_ (h_p)
-------------------------------
%      FLUX  RXN ID    REACTION
---  ------  --------  --------------------------------------------------
47%  16.3    NADH11    5.0 h_c + nadh_c + q_c --> 4.0 h_p + nad_c + qh2_c
23%   8.16   CYO_2p    2.0 focytc_c + 4.0 h_c + 0.5 o2_c --> 2.0 ficyt...
23%   8.16   MEOHDH    2.0 ficytc_c + meoh_p --> fald_p + 2.0 focytc_c...
6%    2.03   CYOO_4p   4.0 h_c + 0.5 o2_c + qh2_c --> h2o_c + 4.0 h_p ...
1%    0.374  PPAppt    h2o_c + ppi_c --> h_p + 2.0 pi_c

CONSUMING REACTIONS -- H_ (h_p)
-------------------------------
%      FLUX  RXN ID    REACTION
---  ------  --------  --------------------------------------------------
99%  34.7    ATPS4rpp  adp_c + 4.0 h_p + pi_c --> atp_c + h2o_c + 3.0 h_c
1%    0.424  NH4tpp    h_p + nh4_p <=> h_c + nh4_c


In [520]:
[rxn for rxn in model.metabolites.h_p.reactions if (model.metabolites.h_p in rxn.reactants or rxn.reversibility) and abs(rxn.flux) > 0]

[<Reaction Htex at 0x7fd7d8442470>,
 <Reaction PIt2rpp at 0x7fd7d84ed1d0>,
 <Reaction FE2t2pp at 0x7fd7d8346630>,
 <Reaction NH4tpp at 0x7fd7d842a278>,
 <Reaction ATPS4rpp at 0x7fd7d8453710>,
 <Reaction MNt2pp at 0x7fd7d8357f28>]

In [522]:
model.metabolites.h_e.reactions

frozenset({<Reaction EX_h_e at 0x7fd7d86cab38>,
           <Reaction Htex at 0x7fd7d8442470>})

In [523]:
for reaction in bigg_OB3b.metabolites.h_p.reactions:
    print(reaction.id, reaction.name, reaction.reaction)

PPAt7 propanoate_symport__periplasm_to_cytoplasm_ h_p + ppa_p --> h_c + ppa_c
Kt3pp potassium_proton_antiporter h_p + k_c --> h_c + k_p
Htex H__transport_via_diffusion__extracellular_to_periplasm_ h_e <=> h_p
PMMODCipp particulate Methane Monooxygenase - Direct Coupling ch4_p + 2.0 focytc_c + 2.0 h_p + o2_p --> 2.0 ficytc_c + h2o_p + meoh_p
L_LACt2rpp _S__lactate_transport__periplasm_to_cytoplasm_ h_p + lac_L_p --> h_c + lac_L_c
PPDODH 1_2_propanediol_dehydrogenase 12ppd_S_p + 4.0 ficytc_c --> 4.0 focytc_c + 2.0 h_p + lald_L_p
UQCYOR_2p ubiquinol_cytochrome_c_reductase 2.0 ficytc_c + qh2_c --> 2.0 focytc_c + 2.0 h_p + q_c
ACACt2pp acetoacetate_symport__periplasm_to_cytoplasm_ acac_p + h_p --> acac_c + h_c
PYRt2rpp pyruvate_symport__periplasm_to_cytoplasm_ h_p + pyr_p --> h_c + pyr_c
CYO_2p cytochrome_c_oxidase 2.0 focytc_c + 4.0 h_c + 0.5 o2_c --> 2.0 ficytc_c + h2o_c + 2.0 h_p
ETOHDH quinoprotein_ethanol_dehydrogenase etoh_p + 2.0 ficytc_c <=> acald_p + 2.0 focytc_c + 2.0 h_p
PIt2rpp 

Inspect whether the following reactions pump H+ out of the system against concentration gradient:

PPDODH PPDODH: 12ppd_S_p + 4.0 ficytc_c --> 4.0 focytc_c + 2.0 h_p + lald_L_p --> mediated by cytochrome C
ETOHDH ETOHDH: etoh_p + 2.0 ficytc_c <=> acald_p + 2.0 focytc_c + 2.0 h_p --> mediated by cytochrome C
MEOHDH MEOHDH: 2.0 ficytc_c + meoh_p --> fald_p + 2.0 focytc_c + 2.0 h_p --> mediated by cytochrome C
NADH11 NADH11: 5.0 h_c + nadh_c + q_c --> 4.0 h_p + nad_c + qh2_c --> energy cost, part of electron-transport chain 
NH4tpp NH4tpp: h_p + nh4_p <=> h_c + nh4_c --> type of transporter
FALDHp FALDHp: fald_p + 2.0 ficytc_c + h2o_p --> 2.0 focytc_c + for_p + 3.0 h_p --> mediated by cytochrome C
PPAppt PPAppt: h2o_c + ppi_c --> h_p + 2.0 pi_c --> energy cost

In [525]:
h_dubious_reactions = ['PPDODH', 'ETOHDH', 'MEOHDH', 'NADH11', 'NH4tpp', 'FALDHp', 'PPAppt']

In [527]:
for reaction in h_dubious_reactions:
    print(model.reactions.get_by_id(reaction).name, model.reactions.get_by_id(reaction))

1_2_propanediol_dehydrogenase PPDODH: 12ppd_S_p + 4.0 ficytc_c --> 4.0 focytc_c + 2.0 h_p + lald_L_p
quinoprotein_ethanol_dehydrogenase ETOHDH: etoh_p + 2.0 ficytc_c <=> acald_p + 2.0 focytc_c + 2.0 h_p
methanol_dehydrogenase MEOHDH: 2.0 ficytc_c + meoh_p --> fald_p + 2.0 focytc_c + 2.0 h_p
NADH_ubiquinone_oxidoreductase_I NADH11: 5.0 h_c + nadh_c + q_c --> 4.0 h_p + nad_c + qh2_c
NH3_transport__periplasm_to_cytoplasm_ NH4tpp: h_p + nh4_p <=> h_c + nh4_c
formaldehyde_dehydrogenase FALDHp: fald_p + 2.0 ficytc_c + h2o_p --> 2.0 focytc_c + for_p + 3.0 h_p
H____translocating_pyrophosphatases PPAppt: h2o_c + ppi_c --> h_p + 2.0 pi_c


In [528]:
model.reactions.PPDODH

<Reaction PPDODH at 0x7fd7d85624a8>

In [545]:
print(model.reactions.NH4tpp)
model.reactions.NH4tpp.bounds = (0, 1000)

NH4tpp: h_p + nh4_p --> h_c + nh4_c


In [530]:
model_Bath.reactions.NH4tpp

<Reaction NH4tpp at 0x7fd7df210128>

In [531]:
model.reactions.FALDHp

<Reaction FALDHp at 0x7fd7d8574e80>

Double checking H+ proton leakage outside of GEM:

In [538]:
[rxn for rxn in model.metabolites.h_p.reactions if (model.metabolites.h_p in rxn.reactants and rxn.reversibility) and abs(rxn.flux) > 0]

[]

In [539]:
[rxn for rxn in model.metabolites.h_p.reactions if (model.metabolites.h_p in rxn.products and rxn.reversibility) and abs(rxn.flux) > 0]

[<Reaction Htex at 0x7fd7d8442470>]

In [547]:
print(model.reactions.Htex)
model.reactions.Htex.bounds = (0, 1000)
print(model.reactions.Htex)

Htex: h_e --> h_p
Htex: h_e --> h_p


In [550]:
h_p_reactants_irr = [rxn for rxn in model.metabolites.h_p.reactions if (model.metabolites.h_p in rxn.reactants) and abs(rxn.flux) > 0]

In [551]:
for reaction in h_p_reactants_irr:
    print(reaction.name,  reaction)

PiT_phosphate_transporter PIt2rpp: h_p + pi_p --> h_c + pi_c
iron_proton_symport FE2t2pp: fe2_p + h_p --> fe2_c + h_c
NH3_transport__periplasm_to_cytoplasm_ NH4tpp: h_p + nh4_p --> h_c + nh4_c
ATP_synthase ATPS4rpp: adp_c + 4.0 h_p + pi_c --> atp_c + h2o_c + 3.0 h_c
manganese_proton_symport MNt2pp: h_p + mn2_p --> h_c + mn2_c


In [552]:
h_p_products_irr = [rxn for rxn in model.metabolites.h_p.reactions if (model.metabolites.h_p in rxn.products) and abs(rxn.flux) > 0]
for reaction in h_p_products_irr:
    print(reaction.name,  reaction)

H__transport_via_diffusion__extracellular_to_periplasm_ Htex: h_e --> h_p
cytochrome_c_oxidase CYO_2p: 2.0 focytc_c + 4.0 h_c + 0.5 o2_c --> 2.0 ficytc_c + h2o_c + 2.0 h_p
cytochrome_o_ubiquinol_oxidase CYOO_4p: 4.0 h_c + 0.5 o2_c + qh2_c --> h2o_c + 4.0 h_p + q_c
methanol_dehydrogenase MEOHDH: 2.0 ficytc_c + meoh_p --> fald_p + 2.0 focytc_c + 2.0 h_p
NADH_ubiquinone_oxidoreductase_I NADH11: 5.0 h_c + nadh_c + q_c --> 4.0 h_p + nad_c + qh2_c
H____translocating_pyrophosphatases PPAppt: h2o_c + ppi_c --> h_p + 2.0 pi_c


#### NH3 
Following 8. Corrections to Validation Methylococcus capsulatus GEM complimentary scripts:

In [574]:
[rxn for rxn in model.metabolites.nh4_c.reactions if rxn.reversibility or model.metabolites.nh4_c in rxn.reactants]

[<Reaction Biomass_Mextorquens_AM1 at 0x7fd7d847ec50>,
 <Reaction PYAM5POr at 0x7fd7d87324a8>,
 <Reaction GLNS at 0x7fd7d8465dd8>,
 <Reaction GMPS at 0x7fd7d87721d0>,
 <Reaction GLUDy at 0x7fd7d87fde48>,
 <Reaction HMBS at 0x7fd7d83f0a58>,
 <Reaction CYSTL at 0x7fd7d83f0e80>,
 <Reaction NNAMr at 0x7fd7d83cb320>,
 <Reaction NADS1 at 0x7fd7d86dafd0>,
 <Reaction Biomass_Mextorquens_AM1_core at 0x7fd7d81c2be0>]

In [555]:
model.metabolites.nh4_c.summary()

PRODUCING REACTIONS -- NH3 (nh4_c)
----------------------------------
%       FLUX  RXN ID    REACTION
---  -------  --------  --------------------------------------------------
98%  0.424    NH4tpp    h_p + nh4_p --> h_c + nh4_c
2%   0.00904  THRD_L    thr_L_c --> 2obut_c + nh4_c

CONSUMING REACTIONS -- NH3 (nh4_c)
----------------------------------
%       FLUX  RXN ID    REACTION
---  -------  --------  --------------------------------------------------
82%  0.353    GLUDy     akg_c + h_c + nadph_c + nh4_c <=> glu_L_c + h2o...
17%  0.0747   GLNS      atp_c + glu_L_c + nh4_c --> adp_c + gln_L_c + h...
1%   0.00526  GMPS      atp_c + nh4_c + xmp_c --> amp_c + gmp_c + 2.0 h...


In [565]:
model_Bath.metabolites.nh4_c.summary()

PRODUCING REACTIONS -- Ammonium (nh4_c)
---------------------------------------
%       FLUX  RXN ID    REACTION
----  ------  --------  --------------------------------------------------
97%   1.46    NTRIR2y   5.0 h_c + 3.0 nadph_c + no2_c --> 2.0 h2o_c + 3...
3%    0.0434  THRD_L    thr__L_c --> 2obut_c + nh4_c

CONSUMING REACTIONS -- Ammonium (nh4_c)
---------------------------------------
%       FLUX  RXN ID    REACTION
----  ------  --------  --------------------------------------------------
100%  1.5     ALAD_L    h_c + nadh_c + nh4_c + pyr_c --> ala__L_c + h2o...


In [559]:
reaction_id_search(model, "alani")

UAMAS UDP_N_acetylmuramate_alanine_ligase ala_L_c + atp_c + uamr_c --> adp_c + h_c + pi_c + uama_c CQW49_RS11265
PHETRS phenylalanine_tRNA_synthetase atp_c + phe_L_c + trnaphe_c --> amp_c + phetrna_c + ppi_c CQW49_RS15785 and CQW49_RS15795
ALATA_D D_alanine_aminotransferase glu_D_c + pyr_c <=> akg_c + ala_D_c META1_0774
ALAR alanine_racemase ala_L_c <=> ala_D_c CQW49_RS10820 or CQW49_RS10555
ANACMURLALAA anhydro_N_acetylmuramoyl_L_alanine_amidase GlcNacanhMurNacAlagGlucDAPAla_c + h2o_c --> GlcNacanhMurNac_c + alagDgluDAPDala_c CQW49_RS11810
UAMAGS UDP_N_acetylmuramoylalanine__D_glutamate_ligase atp_c + glu_D_c + uama_c --> adp_c + h_c + pi_c + uamag_c CQW49_RS11250
UGMDDS UDP_N_acetylmuramoylalanyl_D_glutamyl_2_6_diaminopimelate_D_alanyl_D_alanine_ligase alaala_c + atp_c + ugmd_c --> adp_c + h_c + pi_c + ugmda_c CQW49_RS11240
ALAALAr D_alanine_D_alanine_ligase_B 2.0 ala_D_c + atp_c --> adp_c + alaala_c + h_c + pi_c CQW49_RS11285


Corrresponding glutamine synthase BiGG-reaction: atp_c + glu__L_c + nh4_c ⇌ adp_c + gln__L_c + h_c + pi_c
Reference: http://bigg.ucsd.edu/universal/reactions/GLNS

Corresponding glutamate synthase BiGG-reaction: akg_h + gln__L_h + h_h + nadh_h ⇌ 2.0 glu__L_h + nad_h
Reference: http://bigg.ucsd.edu/universal/reactions/GLUS


Glutamine synthetase and glutamine-dependent NAD-synthetase are present in OB3b-GEM. However, I would care only about GLNS reaction as it assimilates NH4 (ammonia) well.

In [575]:
reaction_id_search(model, "glutamine")

NADS2 glutamine_dependent_NAD____synthetase atp_c + dnad_c + gln_L_c + h2o_c --> amp_c + glu_L_c + h_c + nad_c + ppi_c CQW49_RS00090
GF6PTAr L_glutamine_D_fructose_6_phosphate_aminotransferase f6p_c + gln_L_c --> gam6p_c + glu_L_c CQW49_RS17260 or CQW49_RS06615
GLNS glutamine_synthetase atp_c + glu_L_c + nh4_c --> adp_c + gln_L_c + h_c + pi_c CQW49_RS20750 or CQW49_RS20740


Glutamate synthase is present in the GEM too:

In [576]:
reaction_id_search(model, "glutamate_s")

GLUDy glutamate_synthase_1 akg_c + h_c + nadph_c + nh4_c <=> glu_L_c + h2o_c + nadp_c CQW49_RS02665 and CQW49_RS02785
FPGS8 folylpolyglutamate_synthase_2 10fthf5glu_c + atp_c + glu_L_c --> 10fthf6glu_c + adp_c + h_c + pi_c CQW49_RS15465
MFPGS5 folylpolyglutamate_synthase_1 atp_c + glu_L_c + mlthf5glu_c --> adp_c + h_c + mlthf6glu_c + pi_c CQW49_RS15465
FPGS10 folylpolyglutamate_synthase_3 atp_c + glu_L_c + thf5glu_c --> adp_c + h_c + pi_c + thf6glu_c CQW49_RS15465
ACGS N_acetylglutamate_synthase accoa_c + glu_L_c <=> acglu_c + coa_c + h_c CQW49_RS12750
GLUSy glutamate_synthase_2 akg_c + gln_L_c + h_c + nadph_c --> 2.0 glu_L_c + nadp_c CQW49_RS02665 and CQW49_RS02785


Let's fix ammonia assimilation in OB3b GEM --> No activities of alanine or glutamate dehydrogenases were detected in cell extracts of M. trichosporium OB3b grown on any source of nitrogen (Shishkina and Trotsenko, 1979; Murrell and Dalton, 1983a,b). **It has been concluded that ammonia was assimilated exclusively via the glutamine synthetase/glutamate synthase pathway (Murrell and Dalton, 1983b)** --> 
Reference: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3615186/#B76

In [583]:
nh4_reactions = [rxn for rxn in model.metabolites.nh4_c.reactions if rxn.reversibility or model.metabolites.nh4_c in rxn.reactants]

In [578]:
model.reactions.GLUDy

<Reaction GLUDy at 0x7fd7d87fde48>

In [581]:
model_Bath.reactions.GLUDxi

<Reaction GLUDxi at 0x7fd7f3ea20b8>

In [589]:
for reaction in nh4_reactions:
    if "Biomass" in reaction.id:
        pass
    else:
        print(reaction.name, reaction)

pyridoxamine_5__phosphate_oxidase PYAM5POr: h2o_c + o2_c + pyam5p_c <=> h2o2_c + nh4_c + pydx5p_c
glutamine_synthetase GLNS: atp_c + glu_L_c + nh4_c --> adp_c + gln_L_c + h_c + pi_c
GMP_synthetase__ammonia_dependent_ GMPS: atp_c + nh4_c + xmp_c --> amp_c + gmp_c + 2.0 h_c + ppi_c
glutamate_synthase_1 GLUDy: akg_c + h_c + nadph_c + nh4_c <=> glu_L_c + h2o_c + nadp_c
porphobilinogen_deaminase HMBS: h2o_c + 4.0 ppbng_c <=> hmbil_c + 4.0 nh4_c
cystathionine_beta_lyase CYSTL: cyst_L_c + h2o_c <=> hcys_L_c + nh4_c + pyr_c
nicotinamidase NNAMr: h2o_c + ncam_c <=> nac_c + nh4_c
NH3_dependent_NAD____synthetase NADS1: atp_c + dnad_c + nh4_c --> amp_c + h_c + nad_c + ppi_c


Reactions that should be fixed so that ammonia assimilation is done via GLNS/GLUDy complex only 

pyridoxamine_5__phosphate_oxidase PYAM5POr: h2o_c + o2_c + pyam5p_c <=> h2o2_c + nh4_c + pydx5p_c --> fixed
glutamine_synthetase GLNS: atp_c + glu_L_c + nh4_c --> adp_c + gln_L_c + h_c + pi_c --> GLNS complex
GMP_synthetase__ammonia_dependent_ GMPS: atp_c + nh4_c + xmp_c --> amp_c + gmp_c + 2.0 h_c + ppi_c --> fixed
glutamate_synthase_1 GLUDy: akg_c + h_c + nadph_c + nh4_c <=> glu_L_c + h2o_c + nadp_c --> fixed directionality based on model_Bath
porphobilinogen_deaminase HMBS: h2o_c + 4.0 ppbng_c <=> hmbil_c + 4.0 nh4_c --> fixed
cystathionine_beta_lyase CYSTL: cyst_L_c + h2o_c <=> hcys_L_c + nh4_c + pyr_c --> fixed
nicotinamidase NNAMr: h2o_c + ncam_c <=> nac_c + nh4_c --> fixed
NH3_dependent_NAD____synthetase NADS1: atp_c + dnad_c + nh4_c --> amp_c + h_c + nad_c + ppi_c


In [597]:
reaction = bigg_OB3b.reactions.PYAM5POr
print(reaction)
with bigg_OB3b as model:
    reaction.bounds = (0, 1000)
    print(model.optimize())
reaction.bounds = (0, 1000)
print(reaction)

PYAM5POr: h2o_c + o2_c + pyam5p_c --> h2o2_c + nh4_c + pydx5p_c
<Solution 0.043 at 0x7fd7ce104390>
PYAM5POr: h2o_c + o2_c + pyam5p_c --> h2o2_c + nh4_c + pydx5p_c


In [595]:
reaction = bigg_OB3b.reactions.GMPS
print(reaction)
with bigg_OB3b as model:
    reaction.bounds = (-1000, 0)
    print(model.optimize())
reaction.bounds = (-1000, 0)
print(reaction)

GMPS: atp_c + nh4_c + xmp_c --> amp_c + gmp_c + 2.0 h_c + ppi_c
<Solution 0.043 at 0x7fd7d1e9a2b0>
GMPS: atp_c + nh4_c + xmp_c <-- amp_c + gmp_c + 2.0 h_c + ppi_c


In [602]:
print(model.reactions.GLUDy)
model.reactions.GLUDy.bounds = (0, 1000)
print(model.reactions.GLUDy)

GLUDy: akg_c + h_c + nadph_c + nh4_c --> glu_L_c + h2o_c + nadp_c
GLUDy: akg_c + h_c + nadph_c + nh4_c --> glu_L_c + h2o_c + nadp_c


In [599]:
model_Bath.reactions.GLUDy

<Reaction GLUDy at 0x7fd7f3e9fa90>

In [605]:
reaction = bigg_OB3b.reactions.HMBS
print(reaction)
with bigg_OB3b as model:
    reaction.bounds = (0, 1000)
    print(model.optimize())
reaction.bounds = (0, 1000)
print(reaction)

HMBS: h2o_c + 4.0 ppbng_c <=> hmbil_c + 4.0 nh4_c
<Solution 0.043 at 0x7fd7cb8024a8>
HMBS: h2o_c + 4.0 ppbng_c --> hmbil_c + 4.0 nh4_c


In [607]:
reaction = bigg_OB3b.reactions.CYSTL
print(reaction)
with bigg_OB3b as model:
    reaction.bounds = (0, 1000)
    print(model.optimize())
reaction.bounds = (0, 1000)
print(reaction)

CYSTL: cyst_L_c + h2o_c <=> hcys_L_c + nh4_c + pyr_c
<Solution 0.043 at 0x7fd7cacfecf8>
CYSTL: cyst_L_c + h2o_c --> hcys_L_c + nh4_c + pyr_c


In [609]:
reaction = bigg_OB3b.reactions.NNAMr
print(reaction)
with bigg_OB3b as model:
    reaction.bounds = (0, 1000)
    print(model.optimize())
reaction.bounds = (0, 1000)
print(reaction)

NNAMr: h2o_c + ncam_c <=> nac_c + nh4_c
<Solution 0.043 at 0x7fd7cacfeb70>
NNAMr: h2o_c + ncam_c --> nac_c + nh4_c


In [612]:
reaction = bigg_OB3b.reactions.NADS1
print(reaction)
with bigg_OB3b as model:
    reaction.bounds = (-1000, 0)
    print(model.optimize())
reaction.bounds = (-1000, 0)
print(reaction)

NADS1: atp_c + dnad_c + nh4_c --> amp_c + h_c + nad_c + ppi_c
<Solution 0.043 at 0x7fd7cdd3e898>
NADS1: atp_c + dnad_c + nh4_c <-- amp_c + h_c + nad_c + ppi_c


Recheck NH3 sanity again:

In [613]:
[rxn for rxn in model.metabolites.nh4_c.reactions if rxn.reversibility or model.metabolites.nh4_c in rxn.reactants]

[<Reaction Biomass_Mextorquens_AM1 at 0x7fd7d847ec50>,
 <Reaction GLNS at 0x7fd7d8465dd8>,
 <Reaction GMPS at 0x7fd7d87721d0>,
 <Reaction GLUDy at 0x7fd7d87fde48>,
 <Reaction NADS1 at 0x7fd7d86dafd0>,
 <Reaction Biomass_Mextorquens_AM1_core at 0x7fd7d81c2be0>]

In [618]:
model.reactions.GMPS.reaction

'atp_c + nh4_c + xmp_c <-- amp_c + gmp_c + 2.0 h_c + ppi_c'

In [619]:
model.reactions.NADS1.reaction

'atp_c + dnad_c + nh4_c <-- amp_c + h_c + nad_c + ppi_c'

In [620]:
model.reactions.GLUDy.reaction

'akg_c + h_c + nadph_c + nh4_c --> glu_L_c + h2o_c + nadp_c'

In [621]:
model.reactions.GLNS.reaction

'atp_c + glu_L_c + nh4_c --> adp_c + gln_L_c + h_c + pi_c'

In [789]:
bigg_OB3b.metabolites.nh4_c.summary()

PRODUCING REACTIONS -- NH3 (nh4_c)
----------------------------------
%       FLUX  RXN ID    REACTION
---  -------  --------  --------------------------------------------------
98%  0.424    NH4tpp    h_p + nh4_p --> h_c + nh4_c
2%   0.00904  THRD_L    thr_L_c --> 2obut_c + nh4_c

CONSUMING REACTIONS -- NH3 (nh4_c)
----------------------------------
%       FLUX  RXN ID    REACTION
---  -------  --------  --------------------------------------------------
82%  0.353    GLUDy     akg_c + h_c + nadph_c + nh4_c --> glu_L_c + h2o...
19%  0.0802   GLNS      atp_c + glu_L_c + nh4_c --> adp_c + gln_L_c + h...


#### NADH
Following 8. Corrections to Validation Methylococcus capsulatus GEM complimentary scripts:

In [627]:
nadh_reactions = [rxn for rxn in model.metabolites.nadh_c.reactions if rxn.reversibility or model.metabolites.nadh_c in rxn.products]

In [626]:
model.metabolites.nadh_c.summary()

PRODUCING REACTIONS -- NADH (nadh_c)
------------------------------------
%      FLUX  RXN ID    REACTION
---  ------  --------  --------------------------------------------------
54%  3.23    FDHr      for_c + nad_c <=> co2_c + nadh_c
40%  2.4     ALDD1     fald_c + h2o_c + nad_c --> for_c + 2.0 h_c + na...
4%   0.244   BDH       bhb_c + nad_c <=> acac_c + h_c + nadh_c
1%   0.0782  PGCDr     3pg_c + nad_c <=> 3php_c + h_c + nadh_c

CONSUMING REACTIONS -- NADH (nadh_c)
------------------------------------
%      FLUX  RXN ID    REACTION
---  ------  --------  --------------------------------------------------
68%  4.08    NADH11    5.0 h_c + nadh_c + q_c --> 4.0 h_p + nad_c + qh2_c
15%  0.887   GLYD      h_c + hpyr_c + nadh_c <=> glyc_R_c + nad_c
9%   0.512   HACD1     3hbcoa_c + nad_c <=> aacoa_c + h_c + nadh_c
6%   0.344   MDH       mal_L_c + nad_c <=> h_c + nadh_c + oaa_c
3%   0.16    GAPD      g3p_c + nad_c + pi_c <=> 13dpg_c + h_c + nadh_c


In [629]:
for reaction in nadh_reactions:
    print(reaction.name, reaction)

3_hydroxyacyl_CoA_dehydrogenase__n_16_0_ HACD7: 3hhdcoa_c + nad_c <=> 3ohdcoa_c + h_c + nadh_c
3_hydroxyacyl_CoA_dehydrogenase__n_18_1_ HACOADH181: 3hcoctd11ecoa_c + nad_c <=> 3ococtd11ecoa_c + h_c + nadh_c
2_hydroxy_3_oxopropionate_reductase__NAD_ TRSARr: glyc_R_c + nad_c <=> 2h3oppan_c + h_c + nadh_c
prephenate_dehydrogenase PPND: nad_c + pphn_c --> 34hpp_c + co2_c + nadh_c
erythrose_4_phosphate_dehydrogenase E4PD: e4p_c + h2o_c + nad_c <=> 4per_c + 2.0 h_c + nadh_c
glycolaldehyde_dehydrogenase GCALDDr: gcald_c + h2o_c + nad_c --> glyclt_c + 2.0 h_c + nadh_c
pyruvate_dehydrogenase PDH: coa_c + nad_c + pyr_c --> accoa_c + co2_c + nadh_c
3_hydroxyacyl_CoA_dehydrogenase__n_14_1_ HACOADH141: 3hctd7ecoa_c + nad_c <=> 3octd7ecoa_c + h_c + nadh_c
3_hydroxyacyl_CoA_dehydrogenase__n_18_0_ HACD8i: 3hodcoa_c + nad_c <=> 3oodcoa_c + h_c + nadh_c
NADH_azoreductase MREDR: 2.0 h_c + mred_c + 2.0 nadh_c <=> Nndmetphenda_c + anth_c + 2.0 nad_c
hydroxypyruvate_reductase GLYD: h_c + hpyr_c + nadh_c <=>

In [631]:
model_Bath.metabolites.nadh_c.summary()

PRODUCING REACTIONS -- NADH(2-) (nadh_c)
----------------------------------------
%      FLUX  RXN ID    REACTION
---  ------  --------  --------------------------------------------------
58%  12.3    FDH       for_c + nad_c --> co2_c + nadh_c
41%   8.61   MDH       mal__L_c + nad_c --> h_c + nadh_c + oaa_c
1%    0.218  ICDHxm    icit_c + nad_c --> akg_c + co2_c + nadh_c

CONSUMING REACTIONS -- NADH(2-) (nadh_c)
----------------------------------------
%      FLUX  RXN ID    REACTION
---  ------  --------  --------------------------------------------------
87%  18.5    SMMOi     ch4_c + h_c + nadh_c + o2_c --> h2o_c + meoh_c ...
7%    1.5    ALAD_L    h_c + nadh_c + nh4_c + pyr_c --> ala__L_c + h2o...
7%    1.46   NADH16pp  4.0 h_c + nadh_c + q8_im --> 3.0 h_p + nad_c + ...


In [633]:
[rxn for rxn in model_Bath.metabolites.nadh_c.reactions if rxn.reversibility or model_Bath.metabolites.nadh_c in rxn.products]

[<Reaction GLUDxi at 0x7fd7f3ea20b8>,
 <Reaction GCCc at 0x7fd7f3ea58d0>,
 <Reaction PGCD_GAPFILLING at 0x7fd7de80b908>,
 <Reaction AKGDH at 0x7fd7db143198>,
 <Reaction GLYCL at 0x7fd7f3e9a198>,
 <Reaction SHCHD2 at 0x7fd7e1642978>,
 <Reaction GAPD at 0x7fd7f3ea91d0>,
 <Reaction IMPD at 0x7fd7dea4a9e8>,
 <Reaction G3PD1 at 0x7fd7f3ead208>,
 <Reaction PDH at 0x7fd7de814a58>,
 <Reaction VALDHr at 0x7fd7dec16a58>,
 <Reaction FDH at 0x7fd7f3eb8278>,
 <Reaction NADTRHD at 0x7fd7dc5f0ac8>,
 <Reaction GLYCDx at 0x7fd7f3e9a358>,
 <Reaction ALCD2ir at 0x7fd7db199b70>,
 <Reaction ICDHxm at 0x7fd7dea75b70>,
 <Reaction UDPGD at 0x7fd7e004a390>,
 <Reaction 34DHOXPEGOX at 0x7fd7dfbbf438>,
 <Reaction DHORD_NADi at 0x7fd7df21bcc0>,
 <Reaction ME1 at 0x7fd7deb39cf8>,
 <Reaction PPND at 0x7fd7df012d30>,
 <Reaction IPMD at 0x7fd7dea4a630>,
 <Reaction NAD_H2 at 0x7fd7dc5f0ef0>,
 <Reaction OBDHc at 0x7fd7dccc0f60>,
 <Reaction MDH at 0x7fd7deb39780>,
 <Reaction HISTD_GAPFILLING at 0x7fd7de9d9f98>,
 <Reactio

In [634]:
model.reactions.ALDD1.knock_out()

In [636]:
model.metabolites.nadh_c.summary()

PRODUCING REACTIONS -- NADH (nadh_c)
------------------------------------
%      FLUX  RXN ID    REACTION
---  ------  --------  --------------------------------------------------
57%  3.23    FDHr      for_c + nad_c <=> co2_c + nadh_c
38%  2.15    MLMPTDH   mlh4mpt_c + nad_c --> meh4mpt_c + nadh_c
4%   0.244   BDH       bhb_c + nad_c <=> acac_c + h_c + nadh_c
1%   0.0782  PGCDr     3pg_c + nad_c <=> 3php_c + h_c + nadh_c

CONSUMING REACTIONS -- NADH (nadh_c)
------------------------------------
%      FLUX  RXN ID    REACTION
---  ------  --------  --------------------------------------------------
72%  4.08    NADH11    5.0 h_c + nadh_c + q_c --> 4.0 h_p + nad_c + qh2_c
16%  0.887   GLYD      h_c + hpyr_c + nadh_c <=> glyc_R_c + nad_c
6%   0.344   MDH       mal_L_c + nad_c <=> h_c + nadh_c + oaa_c
4%   0.256   HACD1     3hbcoa_c + nad_c <=> aacoa_c + h_c + nadh_c
3%   0.16    GAPD      g3p_c + nad_c + pi_c <=> 13dpg_c + h_c + nadh_c


In [637]:
model.metabolites.nadph_c.summary()

PRODUCING REACTIONS -- NADPH (nadph_c)
--------------------------------------
%      FLUX  RXN ID    REACTION
---  ------  --------  --------------------------------------------------
93%  1.08    MLMPTDHy  mlh4mpt_c + nadp_c --> meh4mpt_c + nadph_c
4%   0.0498  ICDHyr    icit_c + nadp_c --> akg_c + co2_c + nadph_c
2%   0.0263  MTHFD     methf_c + nadph_c <=> mlthf_c + nadp_c

CONSUMING REACTIONS -- NADPH (nadph_c)
--------------------------------------
%      FLUX  RXN ID    REACTION
---  ------  --------  --------------------------------------------------
31%  0.353   GLUDy     akg_c + h_c + nadph_c + nh4_c --> glu_L_c + h2o...
22%  0.256   AACOAR    aacoa_c + h_c + nadph_c --> nadp_c + r3hbcoa_c
22%  0.256   CCRC      b2coa_c + co2_c + nadph_c <=> 2Sethmalcoa_c + n...
3%   0.036   ASAD      4pasp_c + h_c + nadph_c <=> aspsa_c + nadp_c + ...
3%   0.0354  KARA1     alac_S_c + h_c + nadph_c <=> 23dhmb_c + nadp_c
2%   0.0264  TDPDRR    dtdp4d6dm_c + h_c + nadph_c <=> dtdprmn_c + nadp_c


In [638]:
model_Bath.metabolites.nad_c.summary()

PRODUCING REACTIONS -- NAD(+) (nad_c)
-------------------------------------
%      FLUX  RXN ID    REACTION
---  ------  --------  --------------------------------------------------
86%  18.5    SMMOi     ch4_c + h_c + nadh_c + o2_c --> h2o_c + meoh_c ...
7%    1.5    ALAD_L    h_c + nadh_c + nh4_c + pyr_c --> ala__L_c + h2o...
7%    1.46   NADH16pp  4.0 h_c + nadh_c + q8_im --> 3.0 h_p + nad_c + ...

CONSUMING REACTIONS -- NAD(+) (nad_c)
-------------------------------------
%      FLUX  RXN ID    REACTION
---  ------  --------  --------------------------------------------------
58%  12.3    FDH       for_c + nad_c --> co2_c + nadh_c
40%   8.61   MDH       mal__L_c + nad_c --> h_c + nadh_c + oaa_c
1%    0.218  ICDHxm    icit_c + nad_c --> akg_c + co2_c + nadh_c


"It has been suggested that most of the reducing power required for methane metabolism is produced by formaldehyde oxidation to formate and then to CO2. --> Idea: Check where the reducing power from methane metabolism comes from? You can do simulations for checking this idea
NAD-dependent formate dehydrogenase from M. trichosporium OB3b has been purified and characterized (Jollie and Lipscomb, 1991). The enzyme is composed of four subunit types and contained flavin, iron, and molybdenum (Jollie and Lipscomb, 1991). The genome of M. trichosporium OB3b predicts one NAD-dependent molybdenum-containing formate dehydrogenase encoded by fdsABGCD and an additional single copy of the alpha subunit (fdhA). The two genes fdsA and fdhA share 81% identity. Only one of them, fdsA as well as the rest of the fds cluster genes were expressed in the transcriptome" --> I think there is no way how I can constrain NADH, NADPH metabolism.
<br><br>Reference: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3615186/#B5

Save H_, NH3 and NADH-curated GEM:

In [656]:
cobra.io.save_json_model(model, memote_directory + draft_gems_dir_8 + "bigg_OB3b_26_11_2019_2.json")

In [655]:
cobra.io.write_sbml_model(model, memote_directory + draft_gems_dir_8 + "bigg_OB3b_26_11_2019_2.xml")

In [658]:
model.summary()

IN FLUXES            OUT FLUXES    OBJECTIVES
-------------------  ------------  -----------------------
o2_e       6.38      h2o_e  7.39   Biomass_Mext...  0.0433
ch4_e      4.08      co2_e  2.34
nh4_e      0.424
h_e        0.147
pi_e       0.0303
k_e        0.00799
so4_e      0.00273
mg2_e      0.000357
fe2_e      0.000348
ca2_e      0.000214
cl_e       0.000214
na1_e      0.000176
mobd_e     0.000153
cu2_e      0.000143
mn2_e      0.000143
zn2_e      0.000143
cobalt2_e  9.52e-06


#### Check GEM for sanity based on Palsson protocol + memote internal tests:

This can be done later (as of 26/11/2019)